---
title: "Spatial Long-Read Analysis"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Spatial Long-Read Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

We will load the following packages
```
library(spotglm)
library(ggplot2)
library(dplyr)
library(patchwork)
```
# Overview
In this vignette we will go through a typical analysis using spotGLM on a spatial long-read dataset. The dataset comes from a mouse [olfactory bulb](https://academic.oup.com/nar/article/51/8/e47/7079641). In the dataset, we identify 11 major cell types;astrocytes, endothelial cells, mesenchymal cells, microglia, monocytes, mural cells, oligodendrocytes, neurons (N), olfactory ensheathing cells (OEC), oligodendrocyte progenitor cells (OPC), and red blood cells. We also are given 5 distinct brain regions; the olfactory nerve layer, glomerular layer, outer plexiform layer, mitral layer, and granule layer. We will identify genes that show evidence of isoform switching in a cell type specific manner. 


## Read in Data

We will read in data with the `read_spatial_long_read()` function. The function returns the following
<details>
  <summary>Output</summary>
  + `coords`: Coordinate matrix. Of dimension 918 by 2
+ `niche`: The one hot encoded region matrix for each spot. Of dimension 918 by 5
+ `deconv`: The deconvolution of each spot. of dimension 918 by 11
+ `library size`: The library size of each spot. A vector of length 918
+ `total_gene_expression`: For each gene, a vector of spot gene expression. A matrix of dimension 918 by 643
+ `isoform expression`: A list with each element corresponding to a gene. `isoform expression[[gene]]` is a list of length two that contains the spot level expression of the top two isoforms for that gene. 
</details>

```
data = spotglm::read_spatial_long_read()
#filter out low deconvolutions (likely noise)
data$deconv[data$deconv < 0.05] = 0
data$deconv = t(apply(data$deconv,1,function(x){x/sum(x)}))

```




## Plot Regions 
```
region = apply(data$regions,1,function(x){colnames(data$regions)[which(x == 1)]})
library(ggplot2)

df <- data.frame(
  x = data$coords[, 1],
  y = data$coords[, 2],
  region = region
)

ggplot(df, aes(x = x, y = y, color = region)) +
  geom_point(size = 2) +
  coord_fixed() +
  theme_minimal() +
  labs(title = "Spots by Region", color = "Region")

```



## Model Overview 
We return to the mouse olfactory bulb data.  We restrict our analysis to only those genes with exactly two isoforms whose summed expression comprise at least 20% of all isoforms expressed for that gene.  We further limited to genes with and total expression across all spots > 50 UMIs. In total, 643 genes passed these two criteria. For each gene, we identified cell types where the gene is adequately expressed. A gene $g$ is defined as being adequately expressed within a cell type $t$ if the total proportion of transcripts belonging to gene$g$  in cell type $t$ is greater than 1/60000. The exact proportion was found by fitting SpotGLM with only the intercept term. The exact model is 


$$Y_{s,g} \sim Pois(\exp(\log(L_s) + \sum_{t} \pi_{s,t}\exp(\beta_{0,g}^t)))$$
In this model, $L_s$ represents the library size of spot $s$,$\pi_{s,t}$ represents the deconvolution estimate of cell type $t$ in spot $s$, and $\beta_{0,g}^t$ represents the intercept estimate for cell type $t$. We then focused on these cell types and applied SpotGLM to compare cell type-specific isoform relative expression between all pairs of regions.  \

The SpotGLM and corresponding test are described as follows: \
For a gene $g$ let the isoform relative proportion the isoform of interest in spot $s$  be $\gamma_s^g$. The isoform relative proportion is defined as
$$\mathbb{P}[\text{UMI is of isoform i |UMI is of gene that isoform belongs to}]$$

$\gamma_s^g$ is a weighted average of the isoform relative proportions of the cell types that make up $s$ with cell type $t$'s weight equal to the cell type mean expression of gene $g$, $\mu_{g,t}$ times the proportion of cell type $t$ in spot $s$, $\pi_{s,t}$.Let $\gamma_t^g$ be the isoform relative proportion for cell type $t$. Then 

$$\gamma_s^g = \frac{\sum_{t}\pi_{s,t}\gamma_t^g\mu_{g,t}}{\sum_{t'\in s}\pi_{s,t'}\mu_{g,t'}}$$
At the single cell level, we fit a cell type specific logistic regression model with our covariate matrix ùëã being the one hot encoded region for each cell. That is that 
$$Y_i^g \sim Bin(n_{i}^g, \gamma^g_i)$$
$$\text{logit}(\gamma^g_i) = X_i\beta^{CT(i),g}$$

$CT(i)$ represents the cell type of cell $i$. $Y_i^g$ represents the number of transcripts belong to the isoform of interest. $n_{i}^g$ represents the total number of transcripts belonging to gene $g$. $\beta^{t,g}_r$ represents the log odds of the event that a transcript belonging to gene $g$ belongs to the isoform of interest in cell type $t$ in region $r$.$\beta^{t,g}_r$ can be estimated in a cell type specific manner through spotGLM. We simply have to weight our deconvolution matrix $\pi_{s,t}$ by the corresponding gene weight $\mu_{g,t}$.  To test for cell type specific isoform switching between regions $r$ and $r'$ in cell type $t$, we test the null hypothesis

$$H_0:\beta^{t,g}_r = \beta^{t,g}_{r'}$$
Note that both models require use of the spotglm function `run_model`. 



### Step 1: Get Average Expression of Gene For Each Cell Type (~1 minute)
We first fit the intercept model as described above.We make use of the spotglm function "run_model". We specify our response $y$ which we take to be the gene expression values for a gene $g$. our covariates $X$ is just the intercept matrix. $\lambda$ is our deconvolution which we make sure has row sums that sum to 1. Our offset is the library size of the each spot. The initialization used is TRUE  since we want to initialize all parameters. Our family is "spot poisson" since we are fitting a poisson model. 
```
#get number of genes 
ngenes = ncol(data$total_gene_expression)
#initialize intercept estimates
intercept_estimate = vector("list",ngenes)
names(intercept_estimate) = colnames(data$total_gene_expression)
#iterate over each gene 
for(j in c(1:ngenes)){
  intercept_estimate[[j]]=spotglm::run_model(y = data$total_gene_expression[,j],
                                             X = matrix(1,nrow(data$total_gene_expression),1),
                                             lambda = data$deconv,
                                             offset = log(data$library_size),
                                             initialization = T,
                                             family = "spot poisson")
}
```

### Step 2: SpotGLM to Perform Isoform Differential Expression (~1 minute)

We next fit the binomial spotGLM as described above. We again use the "run model" function. This time however,the family is "spot binomial" since we are running a binomial model. We additionally use two parameters that are generally not used when running spotglm `weights` and `ct_cov_weights`. In a binomial regression model, it is necessary to weight observations by the number of trials used. For each gene this corresponds to the total gene expression vector. Additionally, unlike the spot poisson model where the spot expression was a weighted mean of cell type level expression with the weight corresponding to $\pi_{s,t}$, in the spot binomial model, there is an additional cell type specific weight $\mu_{g,t} = \exp(\beta_{0,g}^t)$. We can include this vector in the `ct_cov_weights` argument. We fit the binomial model below.

```
isoform_DE = vector("list",ngenes)
names(isoform_DE) = colnames(data$total_gene_expression)
for(j in c(1:ngenes)){
  #get gene name 
  gene = names(isoform_DE)[j]
  #get weights 
  weights = data$total_gene_expression[,gene]
  #get covariate weights 
  ct_covariate_weights  = exp(intercept_estimate[[gene]]$beta_est)
  #run model 
  isoform_DE[[j]] = spotglm::run_model(y = data$isoform_expression[[gene]][[1]],
                                       X = data$regions,
                                       lambda = data$deconv,
                                       family = "spot binomial",
                                       weights = weights,
                                       ct_cov_weights = ct_covariate_weights,
                                       initialization = T)
}
```

### Step 3: Downstream Analysis: Contrast Test to Identify Isoform Switching Across Regions
We now test for cell type specific isoform switching across regions. This uses the `compute_contrast_test()` function. We append the matrices outputted from the "compute contrast significance" function and perform a global BH correction. We then print out the genes that show significant isoform switching. 
```
#Store all contrast tests 
all_contrast_tests = list()
cell_types = colnames(data$deconv)
regions = colnames(data$regions)
nregion = length(regions)
counter = 1
#test over all cell types and region pairs 
for(ct in cell_types){
  for(j in c(1:nregion)){
    for(k in c(j:nregion)){
      if(j==k){
        next
      }
      region_A = regions[j]
      region_B = regions[k]
      #compute pvalues
      pvals = suppressWarnings(spotglm::compute_contrast_significance(input_list = isoform_DE,
                                           cell_type = ct,
                                           effect_names = c(region_A,region_B),
                                           beta_name = "beta_est",
                                           covariance_name = "vcov",
                                           sided = 2))
      
      contrast_tests = data.frame(gene = pvals$name, cell_type = ct, 
                                  region_A = region_A, region_B = region_B, 
                                  pval = pvals$pval)
      #add results to list 
      all_contrast_tests[[counter]] = contrast_tests
      counter = counter + 1
    }
  }
}
combined_contrast_tests <- dplyr::bind_rows(all_contrast_tests)%>%
                                                    dplyr::filter(!is.na(pval))

#get significant switches
combined_contrast_tests$qval = p.adjust(combined_contrast_tests$pval,method = "BH")
significant_isoform_switches = combined_contrast_tests%>%filter(qval < 0.1)


```

### Results of Contrast Test as a Table 
```
head(significant_isoform_switches,10)

```

### Example Gene: Ap3s1

Here we show a gene, Ap3s1, that shows evidence of isoform switching between the granule layer and other layers in Neurons. Because Neurons are the only important cell types for this gene, we can safely just plot overall isoform relative proportion and attribute any differences to Neurons. 
```
gene <- "Ap3s1"

# Isoform ratio
isoform_vals <- data$isoform_expression[[gene]][[1]]
gene_total <- data$total_gene_expression[, gene]
ratio <- isoform_vals / gene_total
ratio[!is.finite(ratio)] <- NA

# Region labels
region <- apply(data$regions, 1, function(x) {
  colnames(data$regions)[which(x == 1)]
})

# Binary group
region_group <- ifelse(region == "regionGranule Cell Layer (GCL+RMS)", "Granule Layer", "Other")

# Combined dataframe
df <- data.frame(
  x = data$coords[,1],
  y = data$coords[,2],
  ratio = ratio,
  region = region_group
) %>% filter(!is.na(ratio))

# === Plot 1: Violin + boxplot ===
p1 <- ggplot(df, aes(x = region, y = ratio, fill = region)) +
  geom_violin(trim = FALSE, alpha = 0.6) +
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  theme_minimal() +
  labs(
    title = paste(gene, "Isoform/Gene Ratio"),
    x = "", y = "Ratio"
  ) +
  scale_fill_manual(values = c("Granule Layer" = "tomato", "Other" = "skyblue")) +
  theme(legend.position = "none")

# === Plot 2: Spatial plot - Granule Layer only ===
p2 <- df %>%
  filter(region == "Granule Layer") %>%
  ggplot(aes(x = x, y = y, color = ratio)) +
  geom_point(size = 2) +
  scale_color_viridis_c(na.value = "grey90") +
  coord_fixed() +
  theme_minimal() +
  labs(title = "Granule Layer", x = NULL, y = NULL, color = "Ratio")

# === Plot 3: Spatial plot - Other regions ===
p3 <- df %>%
  filter(region == "Other") %>%
  ggplot(aes(x = x, y = y, color = ratio)) +
  geom_point(size = 2) +
  scale_color_viridis_c(na.value = "grey90") +
  coord_fixed() +
  theme_minimal() +
  labs(title = "Other Regions", x = NULL, y = NULL, color = "Ratio")


p1
p2 + p3



```



