---
title: "Visium Analysis"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Visium_analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

We must load the following packages
```
library(spotglm)
library(ggplot2)
library(patchwork)
library(dplyr)
```
# Overview
In this vignette we will go through a typical analysis using spotGLM on a Visium dataset. The dataset comes from a liver colorectal cancer patient. In the dataset, we identify 6 major cell types, hepatocytes, macrophages/myeloid, fibroblasts/stromal, and two subclones of the tumor found via [clonalscope](https://www.biorxiv.org/content/10.1101/2022.07.05.498882v2). For each spot we summarize its neighborhood/niche by computing a weighted sum of deconvolutions with closer spots having more weight. We will then fit a spotGLM to identify cell type specific niche-differential genes. 


## Load Data

We first load the data. 

# Reading in Data
First we read in the merfish data using the function 'read merfish'. This functions returns a list with the following

<details>
  <summary>Output</summary>
  + `coords`: Coordinate matrix. Of dimension 3826 by 2
+ `niche`: The niche vector for each spot. Of dimension 3826 by 7
+ `deconv`: The deconvolution of each spot. of dimension 3826 by 7
+ `counts`: The counts matrix for 2000 genes. Of dimension 3826 by 2000
+ `library size`: The library size of each spot. A vector of length 3826
</details>

```
data = spotglm::read_visium()
```

## Example of Covariate 
We summarize each spot's niche into a vector of length $T$ where $T$ is the number of cell types in the sample. This covariate is called an effective niche.Without going into too much detail, the larger the covariate, the more the corresponding cell type is in the neighborhood of that spot. For example, we see that the tumor subclone 1 niche covariate is high in regions where the tumor deconvolution is also high. 

```
# Prepare data for plotting
plot_df_deconv <- data.frame(
  x = data$coords[, 1],
  y = data$coords[, 2],
  subclone_1 = data$deconv[, "subclone_1"]
)

plot_df_niche <- data.frame(
  x = data$coords[, 1],
  y = data$coords[, 2],
  subclone_1 = data$niche[, "subclone_1"]
)

# Plot 1: deconv
p1 <- ggplot(plot_df_deconv, aes(x = x, y = y, color = subclone_1)) +
  geom_point(size = 2) +
  scale_color_viridis_c() +
  labs(title = "Deconvolution - Subclone 1", color = "Subclone 1") +
  theme_minimal()

# Plot 2: niche
p2 <- ggplot(plot_df_niche, aes(x = x, y = y, color = subclone_1)) +
  geom_point(size = 2) +
  scale_color_viridis_c() +
  labs(title = "Niche Covariate - Subclone 1", color = "Subclone 1") +
  theme_minimal()

# Display side by side
```
```
p1 + p2

```



## Perform SpotGLM (Approx 10 minutes)
We now fit our spotGLM model. Let $X$ be the the covariate matrix where the $i$th row of X, $X_i$ corresponds to the $T$ dimensional effective niche vector described above. The single cell glm model for each cell type is 
$$Y_{i,g} \sim Pois(\mathbb{E}[Y_{i,g}|X_{i}])$$
$$\log(\mathbb{E}[Y_{i,g}|X_{i}]) = X_i\beta^{CT(i)}_g + \log(L_i)$$
Here $L_i$ is the library size of cell $i$, $CT(i)$ is the cell type assignment of cell $i$ and $\beta^{CT(i)}_g$ is a vector of dimension $T \times 1$ that acts as a weight vector for cells of type $CT(i)$ for gene $g$. If an index of $\beta^{CT(i)}_g$, $\beta^{CT(i)}_{t',g}$ is not equal to 0, then the amount of cell $t'$ in the niche affects the expression of gene $g$ in cells of type $CT(i)$. This is called a (CT(i),t') niche differential gene. \

For a spot $s$, its gene expression is modeled as the sum of the cells that make up the sot and thus can be modeled as 

$$Y_{s,g} \sim Pois(\mathbb{E}[Y_{s,g}|X_{s}])$$
$$\mathbb{E}[Y_{s,g}|X_{i}] = \log(L_s) + \sum_{t} \pi_{s,t}\exp(X_s\beta^{t}_g) $$
Here $L_s$ is the library size of spot $s$ and $\pi_{s,t}$ is the deconvolution estimate for cell type $t$ in spot $s$. This model can be fit using a spotGLM. to do this we use the "run model" function. This function takes several arguments, the most important of which are below
This model can be fit using a `spotGLM`. To do this, we use the `run_model()` function. 

```
num_genes = ncol(data$counts)
res = vector("list",num_genes)
t1 = Sys.time()
#run spotGLM for each gene
for(j in c(1:num_genes)){
  if(j%%500 == 0){
    print(j)
    print(Sys.time() - t1)
  }
  res[[j]] = spotglm::run_model(y = data$counts[,j],X = data$niche, lambda = data$deconv,
                             family = "spot poisson",offset = log(data$library_size),
                             initialization = T,batch_size = 500)
}
names(res) = colnames(data$counts)

```


### Downstream Analysis: Determining Niche Differential Genes 
After fitting the spotGLM we can evaluate which genes as (t,t') niche differential. In this case we get pvalues/qvalues for testing if genes are (fibroblast,subclone 1) niche differential. We can do this using the `compute_significance()` function. We set side = 1 and direction = "pos" so that the niche differential genes correspond to those that are upregulated in fibroblasts when in the presence of tumor subclone 1.
```
#get significant genes
sig_genes = compute_significance(input_list = res,cell_type = "stromal",
                                 effect_name = "subclone_1",
                                 beta_name = "beta_est",
                                 standard_error_name = "stand_err_mat",
                                 sided = 1,direction = "pos")
sig_genes = sig_genes%>%filter(!is.na(pval))
print("The following genes are (fibroblast,tumor subclone 1) niche differential")
print(rownames(sig_genes)[sig_genes$qval < 0.05])
```

### Downstream Analysis: Determining Niche Differential Marker Genes
We can also perform a contrast test to see if a gene is niche-marker gene. A niche marker gene for a cell type $t$ is one in which 
$$\beta^{t}_{t_1,g} \neq \beta^{t}_{t_2,g}$$
that is that we can differentiate whether cell type $t$ is near a niche enriched by cell type $t_1$ or cell type $t_2$ by looking at its gene expression for gene $g$. We can do this using the `compute_contrast_significance()` function. Here we are finding niche marker genes in fibroblasts when near tumor subclone 1 vs tumor subclone 2. We set side = 1 and direction = "pos" so that the genes correspond to those that are upregulated in fibroblasts when in the presence of tumor subclone 1 vs when near tumor subclone 2.



```
#compute contrast test
sig_genes = compute_contrast_significance(input_list = res,
                                    cell_type = "stromal",
                                    effect_names = c("subclone_1","subclone_2"),
                               beta_name = "beta_est",covariance_name = "vcov",
                               sided = 1,direction = "pos")
sig_genes = sig_genes%>%filter(!is.na(pval))
m = paste0("The following genes are niche marker genes for fibroblasts near",
            " tumor sublcone 1 vs tumor subclone 2")
print(m)
print(rownames(sig_genes)[sig_genes$qval < 0.05])
```


### Plots

Note that ISG15 is both a (fibroblast,subclone 1) niche differential gene as well as a niche marker gene. If we plot the expression of ISG15 we see this visually. The fibroblasts in the subclone 2 niche express much less ISG15 than that in the subclone 1 niche. Additionally, the non tumor region ("other) shows very little expression of ISG15. 


```
gene <- "ISG15"

# Expression clipping
expression <- data$counts[, gene]
expression <- pmax(expression, quantile(expression, 0.05))
expression <- pmin(expression, quantile(expression, 0.95))

#deconvolution clipping
deconv_prop = data$deconv[, "stromal"]
deconv_prop = pmax(deconv_prop , quantile(deconv_prop , 0.01))
deconv_prop = pmin(deconv_prop , quantile(deconv_prop , 0.99))

# === Plot 1: ISG15 Expression ===
df1 <- data.frame(
  x = data$coords[, 1],
  y = data$coords[, 2],
  gene = expression
)

p1 <- ggplot(df1, aes(x = x, y = y, color = gene)) +
  geom_point(size = 2) +
  scale_color_viridis_c() +
  coord_fixed() +
  theme_minimal() +
  labs(
    title = paste0(gene, " Expression"),
    x = "X", y = "Y", color = gene
  )

# === Plot 2: Majority subclones ===
majority_subclone <- colnames(data$deconv)[max.col(data$deconv,
                                                   ties.method = "first")]

df2 <- data.frame(
  x = data$coords[, 1],
  y = data$coords[, 2],
  majority = majority_subclone
)

df2$highlight <- case_when(
  df2$majority == "subclone_1" ~ "subclone_1",
  df2$majority == "subclone_2" ~ "subclone_2",
  TRUE ~ "other"
)

p2 <- ggplot(df2, aes(x = x, y = y, color = highlight)) +
  geom_point(size = 2) +
  scale_color_manual(
    values = c(
      "subclone_1" = "red",
      "subclone_2" = "blue",
      "other" = "grey80"
    )
  ) +
  coord_fixed() +
  theme_minimal() +
  labs(title = "Majority Subclone Regions", color = "Majority Subclone")

# === Plot 3: Stromal proportions ===
df3 <- data.frame(
  x = data$coords[, 1],
  y = data$coords[, 2],
  stromal = deconv_prop
)

p3 <- ggplot(df3, aes(x = x, y = y, color = stromal)) +
  geom_point(size = 2) +
  scale_color_viridis_c() +
  coord_fixed() +
  theme_minimal() +
  labs(title = "Stromal Proportion", color = "Stromal")

# === Combine all plots ===
p1
p2
p3

```


Note that COL4A1 is a (fibroblast,subclone 1) niche differential gene but not niche marker gene. If we plot the expression of COL4A1 we see this visually. 

```
gene <- "COL4A1"

# Expression clipping
expression <- data$counts[, gene]
expression <- pmax(expression, quantile(expression, 0.05))
expression <- pmin(expression, quantile(expression, 0.95))

#deconvolution clipping
deconv_prop = data$deconv[, "stromal"]
deconv_prop = pmax(deconv_prop , quantile(deconv_prop , 0.01))
deconv_prop = pmin(deconv_prop , quantile(deconv_prop , 0.99))

# === Plot 1: ISG15 Expression ===
df1 <- data.frame(
  x = data$coords[, 1],
  y = data$coords[, 2],
  gene = expression
)

p1 <- ggplot(df1, aes(x = x, y = y, color = gene)) +
  geom_point(size = 2) +
  scale_color_viridis_c() +
  coord_fixed() +
  theme_minimal() +
  labs(
    title = paste0(gene, " Expression"),
    x = "X", y = "Y", color = gene
  )

# === Plot 2: Majority subclones ===
majority_subclone <- colnames(data$deconv)[max.col(data$deconv,
                                                   ties.method = "first")]

df2 <- data.frame(
  x = data$coords[, 1],
  y = data$coords[, 2],
  majority = majority_subclone
)

df2$highlight <- case_when(
  df2$majority == "subclone_1" ~ "subclone_1",
  df2$majority == "subclone_2" ~ "subclone_2",
  TRUE ~ "other"
)

p2 <- ggplot(df2, aes(x = x, y = y, color = highlight)) +
  geom_point(size = 2) +
  scale_color_manual(
    values = c(
      "subclone_1" = "red",
      "subclone_2" = "blue",
      "other" = "grey80"
    )
  ) +
  coord_fixed() +
  theme_minimal() +
  labs(title = "Majority Subclone Regions", color = "Majority Subclone")

# === Plot 3: Stromal proportions ===
df3 <- data.frame(
  x = data$coords[, 1],
  y = data$coords[, 2],
  stromal = deconv_prop
)

p3 <- ggplot(df3, aes(x = x, y = y, color = stromal)) +
  geom_point(size = 2) +
  scale_color_viridis_c() +
  coord_fixed() +
  theme_minimal() +
  labs(title = "Stromal Proportion", color = "Stromal")

# === Combine all plots ===
p1
p2
p3
```











