<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Spatial Long-Read Isoform Analysis • spotGLM</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Spatial Long-Read Isoform Analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">spotGLM</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Intro_to_SpotGLM.html">Intro to SpotGLM</a></li>
    <li><a class="dropdown-item" href="../articles/Spatial_ATAC_analysis.html">Spatial ATAC Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Spatial_Long_Read_analysis.html">Spatial Long-Read Isoform Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Vignette_VisiumHD_Mouse_Kidney_analysis.html">Visium HD Workflow - SpotGLM w/ SPARROW</a></li>
    <li><a class="dropdown-item" href="../articles/Visium_analysis.html">Visium Niche DE Analysis</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/kaishumason/spotGLM/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Spatial Long-Read Isoform Analysis</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/kaishumason/spotGLM/blob/HEAD/vignettes/Spatial_Long_Read_analysis.Rmd"><code>vignettes/Spatial_Long_Read_analysis.Rmd</code></a></small>
      <div class="d-none name"><code>Spatial_Long_Read_analysis.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This vignette demonstrates a full workflow using SpotGLM for cell
type-specific spatial isoform analysis on a Visium Oxford Nanopore long
read data set. This dataset comes from <a href="https://academic.oup.com/nar/article/51/8/e47/7079641" class="external-link">this study
of the mouse olfactory bulb</a>. In this data set, there are 11 major
cell types, spread across 5 distinct anatomical layers of the olfactory
bulb. Our goal will be to identify genes that show evidence of isoform
switching in a cell type specific manner across the spatial layers.</p>
<div class="section level3">
<h3 id="read-in-the-data">Read in the data<a class="anchor" aria-label="anchor" href="#read-in-the-data"></a>
</h3>
<p>First, load the following packages needed for plotting and data table
manipulation:</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kaishumason/spotGLM">spotGLM</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span> <span class="co"># for plotting</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span> <span class="co"># for manipulating data tables</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span> <span class="co"># for qqplots side-by-side</span></span></code></pre>
<p>Next, read in the olfactory bulb data with the
<code>read_spatial_olfactory_bulb_long_read_data()</code> function. The
function reads from the public repository where this data is stored, and
returns a list with the following fields:</p>
<ul>
<li>
<code>coords</code>: Coordinate matrix. Of dimension 918 by 2</li>
<li>
<code>niche</code>: The one hot encoded region matrix for each spot.
Of dimension 918 by 5</li>
<li>
<code>deconv</code>: The deconvolution of each spot. of dimension
918 by 11</li>
<li>
<code>library size</code>: The library size of each spot. A vector
of length 918</li>
<li>
<code>total_gene_expression</code>: For each gene, a vector of spot
gene expression. A matrix of dimension 918 by 643</li>
<li>
<code>isoform expression</code>: A list with each element
corresponding to a gene. <code>isoform expression[[gene]]</code> is a
list of length two that contains the spot level expression of the top
two isoforms for that gene.</li>
</ul>
<pre><code><span><span class="va">data</span> <span class="op">=</span> <span class="fu">spotGLM</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/spotGLM/man/read_example_spatial_olfactory_bulb_long_read_data.html" class="external-link">read_example_spatial_olfactory_bulb_long_read_data</a></span><span class="op">(</span><span class="op">)</span></span></code></pre>
<p>The deconvolved cell type proportions sometimes have cell types
contributing very small amounts to a spot We denoise by setting the
proportions of these to 0.</p>
<pre><code><span><span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">[</span><span class="va">data</span><span class="op">$</span><span class="va">deconv</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">]</span> <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">deconv</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">deconv</span>,<span class="fl">1</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="va">x</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="initial-data-visualization">Initial data visualization<a class="anchor" aria-label="anchor" href="#initial-data-visualization"></a>
</h3>
<p>First let’s get a feel for this data by plotting spatial maps of the
regions:</p>
<pre><code><span><span class="va">region</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">regions</span>,<span class="fl">1</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">regions</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">x</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span><span class="op">]</span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  region <span class="op">=</span> <span class="va">region</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span><span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">region</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Spots by Region"</span>, color <span class="op">=</span> <span class="st">"Region"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"olfactory_region_map.png"</span>, plot<span class="op">=</span><span class="va">p1</span>, width<span class="op">=</span><span class="fl">8</span>, height<span class="op">=</span><span class="fl">4</span>, dpi<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span></code></pre>
<p><img src="../reference/figures/olfactory_region_map.png" alt="Olfactory bulb spatial map" style="width:80%;"></p>
<p>Also, let’s see how some of the cell types are distributed across the
regions. Let’s try Neurons (“N”), the most common cell type.</p>
<pre><code><span><span class="va">df</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">df</span>, <span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">)</span></span>
<span><span class="va">p1</span><span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">N</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">""</span>, color <span class="op">=</span> <span class="st">"Neuron"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"olfactory_Neuron_spatial.png"</span>, plot<span class="op">=</span><span class="va">p1</span>, width<span class="op">=</span><span class="fl">8</span>, height<span class="op">=</span><span class="fl">4</span>, dpi<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span></code></pre>
<p><img src="../reference/figures/olfactory_Neuron_spatial.png" alt="Olfactory bulb Ap3s1 spatial map" style="width:80%;"></p>
<p>Now, let’s plot the isoform proportions of a specific gene, Ap3s1,
which we know (spoiler alert!) has spatially-variable isoform
expression. We will normalize the isoform-level expression by their
total to get relative expression:</p>
<pre><code><span><span class="va">Ap3s1_isoform1</span><span class="op">=</span><span class="va">data</span><span class="op">$</span><span class="va">isoform_expression</span><span class="op">$</span><span class="va">Ap3s1</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">Ap3s1_isoform2</span><span class="op">=</span><span class="va">data</span><span class="op">$</span><span class="va">isoform_expression</span><span class="op">$</span><span class="va">Ap3s1</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">Ap3s1_isoform1</span> <span class="op">=</span> <span class="va">Ap3s1_isoform1</span><span class="op">/</span><span class="op">(</span><span class="va">Ap3s1_isoform1</span><span class="op">+</span><span class="va">Ap3s1_isoform2</span><span class="op">)</span></span>
<span><span class="va">Ap3s1_isoform2</span> <span class="op">=</span> <span class="va">Ap3s1_isoform2</span><span class="op">/</span><span class="op">(</span><span class="va">Ap3s1_isoform1</span><span class="op">+</span><span class="va">Ap3s1_isoform2</span><span class="op">)</span></span></code></pre>
<p>Plot the spatial map of the isoform relative expression:</p>
<pre><code><span><span class="va">df</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">df</span>, <span class="va">Ap3s1_isoform1</span>, <span class="va">Ap3s1_isoform2</span><span class="op">)</span></span>
<span><span class="va">p1</span><span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">Ap3s1_isoform1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">""</span>, color <span class="op">=</span> <span class="st">"Ap3s1_isoform1"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"olfactory_Ap3s1_spatial_allregions.png"</span>, plot<span class="op">=</span><span class="va">p1</span>, width<span class="op">=</span><span class="fl">8</span>, height<span class="op">=</span><span class="fl">4</span>, dpi<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span></code></pre>
<p><img src="../reference/figures/olfactory_Ap3s1_spatial_allregions.png" alt="Olfactory bulb Ap3s1 spatial map" style="width:80%;"></p>
<p>If you look carefully, you can see that isoform-1 of the above gene
seems to be depleted in the outer-most layer, the olfactory nerve layer.
The change is not obvious visually, is it significant? Does this gene
switch to isoform 2 in the outer-most layer? If it does, such spatial
patterns in isoform usage can simply be attributed to cell type
composition changes, with no cell type-specific isoform switching, or to
the switch in isoform usage within a cell type. The goal of an spotGLM
analysis is to assess the significance of such spatial variations and
dissect what is happening at the cell type level.</p>
</div>
<div class="section level3">
<h3 id="model-overview">Model Overview<a class="anchor" aria-label="anchor" href="#model-overview"></a>
</h3>
<p>For simplicity, we are restricting our analysis to only those genes
with exactly two isoforms whose summed expression comprise at least 20%
of all isoforms expressed for that gene. We further limit to genes with
total expression across all spots greater than 50 UMIs. This is because
for genes with low expression, we wouldn’t have power to detect isoform
switching, and they would inflate the number of tests we would have to
adjust for in the end. In total, 643 genes passed these two
criteria.</p>
<p>Next, for each gene, we identified the cell types where the gene has
sufficient expression. If a gene is not highly expressed in a given cell
type, then it would be pointless to try to detect isoform switching in
that cell type. A gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
is defined as being adequately expressed within a cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
if the total proportion of transcripts mapping to gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
in cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
is greater than 1/60000. This proportion can be estimated by fitting a
spotGLM for this gene, regressing total expression on the cell type
proportions for each spot. The model is as follows:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Y</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi></mrow></msub><mo>∼</mo><mi>P</mi><mi>o</mi><mi>i</mi><mi>s</mi><mrow><mo stretchy="true" form="prefix">(</mo><mo>exp</mo><mrow><mo stretchy="true" form="prefix">(</mo><mo>log</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>L</mi><mi>s</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><munder><mo>∑</mo><mi>t</mi></munder><msub><mi>π</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><mo>exp</mo><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>β</mi><mrow><mn>0</mn><mo>,</mo><mi>g</mi></mrow><mi>t</mi></msubsup><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">Y_{s,g} \sim Pois(\exp(\log(L_s) + \sum_{t} \pi_{s,t}\exp(\beta_{0,g}^t)))</annotation></semantics></math>
In this model,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>L</mi><mi>s</mi></msub><annotation encoding="application/x-tex">L_s</annotation></semantics></math>
is the library size of spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>,<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>π</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">\pi_{s,t}</annotation></semantics></math>
is the deconvolution estimate of cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
in spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>,
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>β</mi><mrow><mn>0</mn><mo>,</mo><mi>g</mi></mrow><mi>t</mi></msubsup><annotation encoding="application/x-tex">\beta_{0,g}^t</annotation></semantics></math>
is the intercept estimate for cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>.</p>
<p>To identify cell type-specific isoform switching, we fit ths spotGLM
model:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Y</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi></mrow></msub><mo>∼</mo><mi>B</mi><mi>i</mi><mi>n</mi><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>n</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi></mrow></msub><mo>,</mo><msub><mi>p</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo></mrow><annotation encoding="application/x-tex">Y_{s,g} \sim Bin(n_{s,g},p_{s,g}),</annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Y</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi></mrow></msub><annotation encoding="application/x-tex">Y_{s,g}</annotation></semantics></math>
is the count of isoform
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>1</mn><annotation encoding="application/x-tex">1</annotation></semantics></math>
of gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
in spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>n</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi></mrow></msub><annotation encoding="application/x-tex">n_{s,g}</annotation></semantics></math>
is the total expression of gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
in spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>,
and
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>p</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi></mrow></msub><mo>=</mo><munder><mo>∑</mo><mi>t</mi></munder><msub><mi>w</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi><mo>,</mo><mi>t</mi></mrow></msub><msup><mtext mathvariant="normal">Logit</mtext><mrow><mi>−</mi><mn>1</mn></mrow></msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>X</mi><mi>s</mi></msub><msub><mi>β</mi><mrow><mi>t</mi><mo>,</mo><mi>g</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo><mspace width="1.0em"></mspace><msub><mi>w</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi><mo>,</mo><mi>t</mi></mrow></msub><mo>=</mo><mfrac><mrow><msub><mi>π</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><msub><mi>μ</mi><mrow><mi>g</mi><mo>,</mo><mi>t</mi></mrow></msub></mrow><mrow><munder><mo>∑</mo><mrow><mi>t</mi><mi>′</mi></mrow></munder><msub><mi>π</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi><mi>′</mi></mrow></msub><msub><mi>μ</mi><mrow><mi>g</mi><mo>,</mo><mi>t</mi><mi>′</mi></mrow></msub></mrow></mfrac><mi>.</mi></mrow><annotation encoding="application/x-tex">p_{s,g} = \sum_{t}w_{s,g,t}\text{Logit}^{-1}(X_s\beta_{t,g}), \quad w_{s,g,t}=\frac{\pi_{s,t}\mu_{g,t}}{\sum_{t'}\pi_{s,t'}\mu_{g,t'}}.</annotation></semantics></math>
In the above formula,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>π</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">\pi_{s,t}</annotation></semantics></math>
is the deconvolution proportion of cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
in spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>μ</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi><mo>,</mo><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">\mu_{s,g,t}</annotation></semantics></math>
is the expected expression of gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
in cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
in spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>
that is estimated in the above spotGLM regression of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Y</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi></mrow></msub><annotation encoding="application/x-tex">Y_{s,g}</annotation></semantics></math>
on the cell type proportions.</p>
<p>Thus, this analysis requires fitting spotGLM two times, once to get
the estimated total expression of each gene originating from each cell
type in each spot, and once more to get the isoform-specific
expressions. Below, we show how this estimation is done in code.</p>
</div>
<div class="section level3">
<h3 id="step-1-estimate-the-cell-type-specific-total-expression-of-each-gene-in-each-spot-1-minute">Step 1: Estimate the cell type-specific total expression of each
gene in each spot (~1 minute)<a class="anchor" aria-label="anchor" href="#step-1-estimate-the-cell-type-specific-total-expression-of-each-gene-in-each-spot-1-minute"></a>
</h3>
<p>We first use `spotGLM::run_model`` for the cell type-specific
estimation of total expression for each gene in each spot. The function
is run separately for each gene, and thus can be parallelized across
genes for large data sets. In our function call, the response
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>y</mi><annotation encoding="application/x-tex">y</annotation></semantics></math>
is set to the total count for a gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
in each spot, and the covariates
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
is just a vector of 1’s (the intercept, since at this step we are not
looking for differential expression across regions).
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>λ</mi><annotation encoding="application/x-tex">\lambda</annotation></semantics></math>
is the deconvolved cell type proportions in each spot. The offset is set
to the library size of the each spot. The initialization is set to TRUE
since we want to initialize all parameters. Our family is “spot poisson”
since we are fitting a poisson model.</p>
<pre><code><span><span class="co">#get number of genes </span></span>
<span><span class="va">ngenes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">total_gene_expression</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#initialize estimates</span></span>
<span><span class="va">intercept_estimate</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="st">"list"</span>,<span class="va">ngenes</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">intercept_estimate</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">total_gene_expression</span><span class="op">)</span></span>
<span><span class="co">#iterate over each gene </span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">ngenes</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">intercept_estimate</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span><span class="op">=</span><span class="fu">spotGLM</span><span class="fu">::</span><span class="fu"><a href="../reference/run_model.html">run_model</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">total_gene_expression</span><span class="op">[</span>,<span class="va">j</span><span class="op">]</span>,</span>
<span>                                             X <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">total_gene_expression</span><span class="op">)</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>                                             lambda <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">deconv</span>,</span>
<span>                                             offset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">library_size</span><span class="op">)</span>,</span>
<span>                                             initialization <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                                             family <span class="op">=</span> <span class="st">"spot poisson"</span>,batch_size <span class="op">=</span> <span class="fl">32</span>, n_epoch <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="step-2-fit-spotglm-model-to-get-estimate-isoform-specific-coefficients-for-each-cell-type-in-each-region-1-minute">Step 2: Fit SpotGLM model to get estimate isoform-specific
coefficients for each cell type in each region (~1 minute)<a class="anchor" aria-label="anchor" href="#step-2-fit-spotglm-model-to-get-estimate-isoform-specific-coefficients-for-each-cell-type-in-each-region-1-minute"></a>
</h3>
<p>Next, we fit the binomial spotGLM as described above. We again use
the <code>run_model</code> function. This time, however, the family is
“spot binomial” since we are running a binomial model. We additionally
use two parameters that are specific to the binomial model:
<code>weights</code> and <code>ct_cov_weights</code>. In a binomial
regression, it is necessary to specify the total number of trials, which
for our case corresponds to the total gene expression vector.
Additionally, unlike the spot poisson model where the spot expression
was a weighted mean of cell type level expression with the weight
corresponding to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>π</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">\pi_{s,t}</annotation></semantics></math>,
in the spot binomial model, there is an additional cell type specific
weight
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>μ</mi><mrow><mi>g</mi><mo>,</mo><mi>t</mi></mrow></msub><mo>=</mo><mo>exp</mo><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>β</mi><mrow><mn>0</mn><mo>,</mo><mi>g</mi></mrow><mi>t</mi></msubsup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\mu_{g,t} = \exp(\beta_{0,g}^t)</annotation></semantics></math>.
This vector needs to be given to the <code>ct_cov_weights</code>
argument. Thus, we fit the binomial model as follows:</p>
<pre><code><span><span class="va">isoform_DE</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="st">"list"</span>,<span class="va">ngenes</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">isoform_DE</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">total_gene_expression</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">ngenes</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">intercept_estimate</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">==</span><span class="fl">0</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw">next</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="co">#get gene name </span></span>
<span>  <span class="va">gene</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">isoform_DE</span><span class="op">)</span><span class="op">[</span><span class="va">j</span><span class="op">]</span></span>
<span>  <span class="co">#get weights </span></span>
<span>  <span class="va">weights</span> <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">total_gene_expression</span><span class="op">[</span>,<span class="va">gene</span><span class="op">]</span></span>
<span>  <span class="co">#get covariate weights </span></span>
<span>  <span class="va">ct_covariate_weights</span>  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">intercept_estimate</span><span class="op">[[</span><span class="va">gene</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">beta_estimate</span><span class="op">)</span></span>
<span>  <span class="co">#run model </span></span>
<span>  <span class="va">isoform_DE</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu">spotGLM</span><span class="fu">::</span><span class="fu"><a href="../reference/run_model.html">run_model</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">isoform_expression</span><span class="op">[[</span><span class="va">gene</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                                       X <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">regions</span>,</span>
<span>                                       lambda <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">deconv</span>,</span>
<span>                                       family <span class="op">=</span> <span class="st">"spot binomial"</span>,</span>
<span>                                       weights <span class="op">=</span> <span class="va">weights</span>,</span>
<span>                                       ct_cov_weights <span class="op">=</span> <span class="va">ct_covariate_weights</span>,</span>
<span>                                       initialization <span class="op">=</span> <span class="cn">T</span>,batch_size <span class="op">=</span> <span class="fl">32</span>, n_epoch <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="step-3-contrast-regions-for-a-given-cell-type-to-detect-spatial-isoform-switching-1-minute">Step 3: Contrast regions for a given cell type to detect spatial
isoform switching (&lt;1 minute)<a class="anchor" aria-label="anchor" href="#step-3-contrast-regions-for-a-given-cell-type-to-detect-spatial-isoform-switching-1-minute"></a>
</h3>
<p>We now test for cell type specific isoform switching across regions.
This uses the <code><a href="../reference/compute_contrast_significance.html">compute_contrast_significance()</a></code> function, for
which you need to input a specific cell type and the regions you want to
compare. Here, we simply loop over all cell types and all region-pairs.
We store all results into a data frame
<code>combined_contrast_tests</code> which can be viewed or filtered to
get all significant genes for a specific region-pairs of interest.</p>
<pre><code><span><span class="va">cell_types</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">)</span></span>
<span><span class="va">regions</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">regions</span><span class="op">)</span></span>
<span><span class="va">nregion</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span></span>
<span><span class="va">counter</span> <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="va">all_contrast_tests</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>  <span class="co"># list to store results </span></span>
<span><span class="co"># loop over all cell types and region pairs </span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">ct</span> <span class="kw">in</span> <span class="va">cell_types</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">nregion</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw">for</span><span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">j</span><span class="op">:</span><span class="va">nregion</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>      <span class="kw">if</span><span class="op">(</span><span class="va">j</span><span class="op">==</span><span class="va">k</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="kw">next</span></span>
<span>      <span class="op">}</span></span>
<span>      <span class="va">region_A</span> <span class="op">=</span> <span class="va">regions</span><span class="op">[</span><span class="va">j</span><span class="op">]</span></span>
<span>      <span class="va">region_B</span> <span class="op">=</span> <span class="va">regions</span><span class="op">[</span><span class="va">k</span><span class="op">]</span></span>
<span>      <span class="co">#compute pvalues</span></span>
<span>      <span class="va">pvals</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu">spotGLM</span><span class="fu">::</span><span class="fu"><a href="../reference/compute_contrast_significance.html">compute_contrast_significance</a></span><span class="op">(</span>input_list <span class="op">=</span> <span class="va">isoform_DE</span>,</span>
<span>                                           cell_type <span class="op">=</span> <span class="va">ct</span>,</span>
<span>                                           effect_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">region_A</span>,<span class="va">region_B</span><span class="op">)</span>,</span>
<span>                                           beta_name <span class="op">=</span> <span class="st">"beta_estimate"</span>,</span>
<span>                                           covariance_name <span class="op">=</span> <span class="st">"vcov"</span>,</span>
<span>                                           sided <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>      </span>
<span>      <span class="va">contrast_tests</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">pvals</span><span class="op">$</span><span class="va">name</span>, cell_type <span class="op">=</span> <span class="va">ct</span>, </span>
<span>                                  region_A <span class="op">=</span> <span class="va">region_A</span>, region_B <span class="op">=</span> <span class="va">region_B</span>, </span>
<span>                                  pval <span class="op">=</span> <span class="va">pvals</span><span class="op">$</span><span class="va">pval</span><span class="op">)</span></span>
<span>      <span class="co">#add results to list </span></span>
<span>      <span class="va">all_contrast_tests</span><span class="op">[[</span><span class="va">counter</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">contrast_tests</span></span>
<span>      <span class="va">counter</span> <span class="op">=</span> <span class="va">counter</span> <span class="op">+</span> <span class="fl">1</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span><span class="va">combined_contrast_tests</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">all_contrast_tests</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                                                    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">pval</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>Compute false discovery q-value using Benjamini-Hochberg:</p>
<pre><code><span><span class="va">combined_contrast_tests</span><span class="op">$</span><span class="va">qval</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">combined_contrast_tests</span><span class="op">$</span><span class="va">pval</span>,method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span></code></pre>
<p>Use 10% false discovery rate as a threshold, get significant cell
type-specific isoform switching events between regions:</p>
<pre><code><span><span class="va">significant_isoform_switches</span> <span class="op">=</span> <span class="va">combined_contrast_tests</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">qval</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre>
<div class="section level4">
<h4 id="example-gene-ap3s1">Example Gene: Ap3s1<a class="anchor" aria-label="anchor" href="#example-gene-ap3s1"></a>
</h4>
<p>Here we show the gene we showed above, Ap3s1, which has significant
isoform switching between the granule layer and other layers in Neurons
(q-value &lt; 0.05 in the table above). Thus, let’s compare the relative
expression of isoform 1 for this gene between the granule layer and the
other layers.</p>
<pre><code><span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="st">"Ap3s1"</span></span>
<span></span>
<span><span class="co"># Isoform ratio</span></span>
<span><span class="va">isoform_vals</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">isoform_expression</span><span class="op">[[</span><span class="va">gene</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">gene_total</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">total_gene_expression</span><span class="op">[</span>, <span class="va">gene</span><span class="op">]</span></span>
<span><span class="va">ratio</span> <span class="op">&lt;-</span> <span class="va">isoform_vals</span> <span class="op">/</span> <span class="va">gene_total</span></span>
<span><span class="va">ratio</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/is.finite.html" class="external-link">is.finite</a></span><span class="op">(</span><span class="va">ratio</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span></span>
<span><span class="co"># Region labels</span></span>
<span><span class="va">region</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">regions</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">regions</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">x</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Binary group</span></span>
<span><span class="va">region_group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">region</span> <span class="op">==</span> <span class="st">"regionGranule Cell Layer (GCL+RMS)"</span>, <span class="st">"Granule Layer"</span>, <span class="st">"Other"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combined dataframe</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>,</span>
<span>  ratio <span class="op">=</span> <span class="va">ratio</span>,</span>
<span>  region <span class="op">=</span> <span class="va">region_group</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">ratio</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># === Plot 1: Violin + boxplot ===</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">region</span>, y <span class="op">=</span> <span class="va">ratio</span>, fill <span class="op">=</span> <span class="va">region</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html" class="external-link">geom_violin</a></span><span class="op">(</span>trim <span class="op">=</span> <span class="cn">FALSE</span>, alpha <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.2</span>, outlier.shape <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">gene</span>, <span class="st">"Isoform/Gene Ratio"</span><span class="op">)</span>,</span>
<span>    x <span class="op">=</span> <span class="st">""</span>, y <span class="op">=</span> <span class="st">"Ratio"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Granule Layer"</span> <span class="op">=</span> <span class="st">"tomato"</span>, <span class="st">"Other"</span> <span class="op">=</span> <span class="st">"skyblue"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"olfactory_Ap3s1_violin.png"</span>, plot<span class="op">=</span><span class="va">p1</span>, width<span class="op">=</span><span class="fl">8</span>, height<span class="op">=</span><span class="fl">4</span>, dpi<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span></code></pre>
<p><img src="../reference/figures/olfactory_Ap3s1_violin.png" alt="Olfactory bulb Ap3s1 spatial map" style="width:80%;"></p>
<p>Now, let’s plot the relative isoform 1 expression for this gene
separately for the Granule Layer and the other layers:</p>
<pre><code><span><span class="co"># === Plot 2: Spatial plot - Granule Layer only ===</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="va">df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">region</span> <span class="op">==</span> <span class="st">"Granule Layer"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">ratio</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span>na.value <span class="op">=</span> <span class="st">"grey90"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Granule Layer"</span>, x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span>, color <span class="op">=</span> <span class="st">"Ratio"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># === Plot 3: Spatial plot - Other regions ===</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="va">df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">region</span> <span class="op">==</span> <span class="st">"Other"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">ratio</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span>na.value <span class="op">=</span> <span class="st">"grey90"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Other Regions"</span>, x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span>, color <span class="op">=</span> <span class="st">"Ratio"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"olfactory_Ap3s1_spatial.png"</span>, plot<span class="op">=</span><span class="va">p2</span><span class="op">+</span><span class="va">p3</span>, width<span class="op">=</span><span class="fl">8</span>, height<span class="op">=</span><span class="fl">4</span>, dpi<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span></code></pre>
<p><img src="../reference/figures/olfactory_Ap3s1_spatial.png" alt="Olfactory bulb Ap3s1 spatial map" style="width:80%;"></p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kaishu Mason, Yijia Jiang, Nancy R. Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
