<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Spatial ATAC Analysis ‚Ä¢ spotGLM</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Spatial ATAC Analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">spotGLM</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Intro_to_SpotGLM.html">Intro to SpotGLM</a></li>
    <li><a class="dropdown-item" href="../articles/Spatial_ATAC_analysis.html">Spatial ATAC Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Spatial_Long_Read_analysis.html">Spatial Long-Read Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Vignette_VisiumHD_Mouse_Kidney_analysis.html">Visium Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Visium_analysis.html">Visium Analysis</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/kaishumason/spotGLM/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Spatial ATAC Analysis</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/kaishumason/spotGLM/blob/HEAD/vignettes/Spatial_ATAC_analysis.Rmd"><code>vignettes/Spatial_ATAC_analysis.Rmd</code></a></small>
      <div class="d-none name"><code>Spatial_ATAC_analysis.Rmd</code></div>
    </div>

    
    
<p>We load the following packages.</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">spotglm</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">sparrow</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span></code></pre>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this vignette we will go through a typical analysis using spotGLM
on a spatial ATAC dataset. The dataset comes from a <a href="https://www.nature.com/articles/s41586-023-05795-1#Fig1" class="external-link">p22 mouse
brain</a>. In the dataset, we identify 13 major cell types, the most
important for this analysis being oligodendrocytes. We are also given 14
distinct regions, the most important of which are the cortex (C0) and
the corpus callosum (C3). From the reads in the dataset, we compute the
motif score by aggregating reads that overlap with various transcription
factor motif signatures. Our goal is to identify changes in motif
signature accessibility across regions in a cell type specific
manner.</p>
<div class="section level3">
<h3 id="read-data">Read Data<a class="anchor" aria-label="anchor" href="#read-data"></a>
</h3>
We will read in data with the <code><a href="../reference/read_spatial_long_read.html">read_spatial_long_read()</a></code>
function. The function returns the following
<details><summary>
Output
</summary><ul>
<li>
<code>coords</code>: Coordinate matrix. Of dimension 9215 by 2</li>
<li>
<code>regions</code>: The one hot encoded region matrix for each
spot. Of dimension 9215 by 14</li>
<li>
<code>deconv</code>: The deconvolution of each spot. of dimension
9215 by 9</li>
<li>
<code>motif_scores</code>: The motif score for each spot. A matrix
of dimension 9215 by 746.</li>
</ul></details><pre><code><span><span class="va">data</span> <span class="op">=</span> <span class="fu">spotglm</span><span class="fu">::</span><span class="fu">read_spatial_atac</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="plot-of-spatial-regions">Plot of Spatial Regions<a class="anchor" aria-label="anchor" href="#plot-of-spatial-regions"></a>
</h3>
<pre><code><span><span class="co">#number of regions</span></span>
<span><span class="va">region</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">regions</span>,<span class="fl">1</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">x</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">n_regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">region</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a data frame for plotting</span></span>
<span><span class="va">plot_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  region <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">region</span><span class="op">)</span>  <span class="co"># convert to factor for discrete coloring</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">region</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://scales.r-lib.org/reference/pal_hue.html" class="external-link">hue_pal</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="va">n_regions</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Spots Colored by Region"</span>, color <span class="op">=</span> <span class="st">"Region"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span></code></pre>
</div>
<div class="section level3">
<h3 id="model">Model<a class="anchor" aria-label="anchor" href="#model"></a>
</h3>
<p>Suppose we have single cell data. Let
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Y</mi><mrow><mi>i</mi><mo>,</mo><mi>m</mi></mrow></msub><annotation encoding="application/x-tex">Y_{i,m}</annotation></semantics></math>
be the motif score of cell
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
for motif
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>.
We model
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Y</mi><mrow><mi>i</mi><mo>,</mo><mi>m</mi></mrow></msub><annotation encoding="application/x-tex">Y_{i,m}</annotation></semantics></math>
as follows
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Y</mi><mrow><mi>i</mi><mo>,</mo><mi>m</mi></mrow></msub><mo>‚àº</mo><mi>ùí©</mi><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>Œ≤</mi><mrow><mi>m</mi><mo>,</mo><msub><mi>r</mi><mi>i</mi></msub></mrow><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msubsup><mo>+</mo><msubsup><mi>Œ≤</mi><mrow><mi>m</mi><mo>,</mo><mn>0</mn></mrow><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msubsup><mo>,</mo><msubsup><mi>œÉ</mi><mi>m</mi><mn>2</mn></msubsup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">Y_{i,m} \sim \mathcal{N}(\beta^{CT(i)}_{m,r_i} + \beta^{CT(i)}_{m,0},\sigma_m^2)</annotation></semantics></math>
Here
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">CT(i)</annotation></semantics></math>
represents the cell type of cell
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>r</mi><mi>i</mi></msub><annotation encoding="application/x-tex">r_i</annotation></semantics></math>
represents the region that cell
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
resides in.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>Œ≤</mi><mrow><mi>m</mi><mo>,</mo><msub><mi>r</mi><mi>i</mi></msub></mrow><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msubsup><annotation encoding="application/x-tex">\beta^{CT(i)}_{m,r_i}</annotation></semantics></math>
represents the expected motif score of a cell of type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">CT(i)</annotation></semantics></math>
in region
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>r</mi><mi>i</mi></msub><annotation encoding="application/x-tex">r_i</annotation></semantics></math>.
For a spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>,
its motif score
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Y</mi><mrow><mi>s</mi><mo>,</mo><mi>m</mi></mrow></msub><annotation encoding="application/x-tex">Y_{s,m}</annotation></semantics></math>
is a weighted average of the motif scores for all cell that reside in
it, i.e
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Y</mi><mrow><mi>s</mi><mo>,</mo><mi>m</mi></mrow></msub><mo>‚àº</mo><mi>ùí©</mi><mrow><mo stretchy="true" form="prefix">(</mo><munder><mo>‚àë</mo><mi>t</mi></munder><msub><mi>œÄ</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>Œ≤</mi><mrow><mi>m</mi><mo>,</mo><msub><mi>r</mi><mi>s</mi></msub></mrow><mi>t</mi></msubsup><mo>+</mo><msubsup><mi>Œ≤</mi><mrow><mi>m</mi><mo>,</mo><mn>0</mn></mrow><mi>t</mi></msubsup><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo><msubsup><mi>œÉ</mi><mi>m</mi><mn>2</mn></msubsup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">Y_{s,m} \sim \mathcal{N}( \sum_{t}\pi_{s,t}(\beta^{t}_{m,r_s} + \beta^{t}_{m,0}),\sigma_m^2)</annotation></semantics></math>
the weights for cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
is the deconvolution amount of cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
in spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>œÄ</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">\pi_{s,t}</annotation></semantics></math>.
This model can be fit with a spotGLM. To test for differences in motif
accessibility in a cell type specific manner, we test
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mn>0</mn></msub><mo>:</mo><msubsup><mi>Œ≤</mi><mrow><mi>m</mi><mo>,</mo><mi>r</mi></mrow><mi>t</mi></msubsup><mo>=</mo><msubsup><mi>Œ≤</mi><mrow><mi>m</mi><mo>,</mo><mi>r</mi><mi>‚Ä≤</mi></mrow><mi>t</mi></msubsup></mrow><annotation encoding="application/x-tex">H_0: \beta^{t}_{m,r} = \beta^{t}_{m,r'}</annotation></semantics></math></p>
<div class="section level4">
<h4 id="spotglm-of-motif-score-on-region-2-minutes">SpotGLM of Motif Score on Region (~2 minutes)<a class="anchor" aria-label="anchor" href="#spotglm-of-motif-score-on-region-2-minutes"></a>
</h4>
<p>We can fit the model described above via the <code>spot_lm</code>
function.</p>
<pre><code><span><span class="va">num_motifs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">motif_scores</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="st">"list"</span>,<span class="va">num_motifs</span><span class="op">)</span></span>
<span></span>
<span><span class="va">t1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">num_motifs</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="va">j</span><span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span><span class="fl">100</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">t1</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">res</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="../reference/spot_lm.html">spot_lm</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">motif_scores</span><span class="op">[</span>,<span class="va">j</span><span class="op">]</span>,X <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">regions</span>,</span>
<span>                     lambda <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">motif_scores</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="oligodendrocyte-differential-motif-accessibility">Oligodendrocyte Differential Motif Accessibility<a class="anchor" aria-label="anchor" href="#oligodendrocyte-differential-motif-accessibility"></a>
</h4>
<p>We can compute pvalues and qvalues for testing for accessibility
differences in oligodendrocytes in the corpus callosum vs the cortex
using the <code><a href="../reference/compute_contrast_significance.html">compute_contrast_significance()</a></code> function.</p>
<pre><code><span><span class="va">sig_genes</span> <span class="op">=</span> <span class="fu"><a href="../reference/compute_contrast_significance.html">compute_contrast_significance</a></span><span class="op">(</span>input_list <span class="op">=</span> <span class="va">res</span>,</span>
<span>                                    cell_type <span class="op">=</span> <span class="st">"Oligodendrocytes"</span>,</span>
<span>                                    effect_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"regionC3"</span>,<span class="st">"regionC0"</span><span class="op">)</span>,</span>
<span>                                    beta_name <span class="op">=</span> <span class="st">"beta_estimate"</span>,covariance_name <span class="op">=</span> <span class="st">"vcov"</span>,</span>
<span>                                    sided <span class="op">=</span> <span class="fl">1</span>,direction <span class="op">=</span> <span class="st">"pos"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sig_genes</span> <span class="op">=</span> <span class="va">sig_genes</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">pval</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Motifs exhibiting differential accessibility in Oligodendrocytes"</span>,</span>
<span>           <span class="st">" in corpus callosum vs cortex"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sig_genes</span><span class="op">)</span><span class="op">[</span><span class="va">sig_genes</span><span class="op">$</span><span class="va">qval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="qqplot">QQplot<a class="anchor" aria-label="anchor" href="#qqplot"></a>
</h4>
<pre><code><span><span class="co">#get pvalues from spotglm </span></span>
<span><span class="va">pvals</span> <span class="op">=</span> <span class="va">sig_genes</span><span class="op">$</span><span class="va">pval</span></span>
<span><span class="va">P</span> <span class="op">=</span> <span class="va">pvals</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">pvals</span><span class="op">)</span> <span class="op">==</span> <span class="cn">F</span><span class="op">]</span></span>
<span><span class="va">P</span><span class="op">[</span><span class="va">P</span><span class="op">&gt;</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="fl">1e-10</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="fl">1</span><span class="op">-</span><span class="fl">1e-10</span></span>
<span><span class="va">P</span><span class="op">[</span><span class="va">P</span><span class="op">&lt;</span><span class="op">(</span><span class="fl">1e-10</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="fl">1e-10</span></span>
<span></span>
<span><span class="co">#get qvalues to color </span></span>
<span><span class="va">pvals_adj</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">pvals</span>,method<span class="op">=</span><span class="st">"BH"</span><span class="op">)</span></span>
<span><span class="co">#convert to z-scores </span></span>
<span><span class="va">zscores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">qnorm</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">P</span><span class="op">)</span>  <span class="co"># Inverse normal transformation</span></span>
<span></span>
<span><span class="co"># Generate QQ-Plot of transformed p-values (Z-Scores)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/qqnorm.html" class="external-link">qqnorm</a></span><span class="op">(</span><span class="va">zscores</span>, main<span class="op">=</span> <span class="st">"QQplot of pvalues"</span>, xlab<span class="op">=</span><span class="st">""</span>, ylab<span class="op">=</span><span class="st">""</span>, axes<span class="op">=</span><span class="cn">FALSE</span>, pch<span class="op">=</span><span class="fl">16</span>, </span>
<span>       col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">pvals_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="st">"red"</span>, <span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="co">#significant ones are red</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/qqnorm.html" class="external-link">qqline</a></span><span class="op">(</span><span class="va">zscores</span>, col<span class="op">=</span><span class="st">"red"</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span>  <span class="co"># Add reference line</span></span>
<span></span>
<span><span class="co"># Add surrounding box for clarity</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/box.html" class="external-link">box</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span></code></pre>
</div>
<div class="section level4">
<h4 id="example-motif-sox4">Example Motif: SOX4<a class="anchor" aria-label="anchor" href="#example-motif-sox4"></a>
</h4>
<p>We showcase an example of a motif that shows differential motif
accessibility in oligodendrocytes between the corpus callosum and the
cortex.</p>
<pre><code><span><span class="va">motif</span> <span class="op">&lt;-</span> <span class="st">"SOX4"</span></span>
<span></span>
<span><span class="co"># Filter spots in regionC3 or regionC0</span></span>
<span><span class="va">region_filter</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">regions</span><span class="op">[</span>, <span class="st">"regionC3"</span><span class="op">]</span> <span class="op">==</span> <span class="fl">1</span> <span class="op">|</span> <span class="va">data</span><span class="op">$</span><span class="va">regions</span><span class="op">[</span>, <span class="st">"regionC0"</span><span class="op">]</span> <span class="op">==</span> <span class="fl">1</span></span>
<span></span>
<span><span class="co"># Motif score with truncation</span></span>
<span><span class="va">Sox_score</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">motif_scores</span><span class="op">[</span><span class="va">region_filter</span>, <span class="va">motif</span><span class="op">]</span></span>
<span><span class="va">Sox_score</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmin</a></span><span class="op">(</span><span class="va">Sox_score</span>, <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">Sox_score</span>, <span class="fl">0.99</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">Sox_score</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmax</a></span><span class="op">(</span><span class="va">Sox_score</span>, <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">Sox_score</span>, <span class="fl">0.01</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Region labels</span></span>
<span><span class="va">region_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">regions</span><span class="op">[</span><span class="va">region_filter</span>, <span class="st">"regionC3"</span><span class="op">]</span> <span class="op">==</span> <span class="fl">1</span>, <span class="st">"Corpus Callosum"</span>, <span class="st">"Cortex"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Oligodendrocyte proportion</span></span>
<span><span class="va">oligo_score</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">[</span><span class="va">region_filter</span>, <span class="st">"Oligodendrocytes"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Create plotting data frame</span></span>
<span><span class="va">plot_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span><span class="va">region_filter</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span><span class="va">region_filter</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  Sox <span class="op">=</span> <span class="va">Sox_score</span>,</span>
<span>  region <span class="op">=</span> <span class="va">region_label</span>,</span>
<span>  oligodendrocytes <span class="op">=</span> <span class="va">oligo_score</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># === Plot 1: SOX4 Motif Scores ===</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">Sox</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">motif</span>, <span class="st">" Motif Scores"</span><span class="op">)</span>, color <span class="op">=</span> <span class="va">motif</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># === Plot 2: Region membership ===</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">region</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cortex"</span> <span class="op">=</span> <span class="st">"#1f78b4"</span>, <span class="st">"Corpus Callosum"</span> <span class="op">=</span> <span class="st">"#33a02c"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Region Membership"</span>, color <span class="op">=</span> <span class="st">"Region"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># === Plot 3: Oligodendrocyte proportion ===</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">oligodendrocytes</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Oligodendrocyte Deconvolution Proportion"</span>, color <span class="op">=</span> <span class="st">"Oligos"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span></span>
<span><span class="va">p2</span></span>
<span><span class="va">p3</span></span></code></pre>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kaishu Mason, Yijia Jiang, Nancy R. Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
