<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Spatial Long-Read Analysis • spotglm</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Spatial Long-Read Analysis">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">spotglm</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Intro_to_SpotGLM.html">Intro to SpotGLM</a>
    </li>
    <li>
      <a href="../articles/Spatial_ATAC_analysis.html">Spatial ATAC Analysis</a>
    </li>
    <li>
      <a href="../articles/Spatial_Long_Read_analysis.html">Spatial Long-Read Analysis</a>
    </li>
    <li>
      <a href="../articles/Visium_analysis.html">Visium Analysis</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Spatial Long-Read Analysis</h1>
            
      

      <div class="hidden name"><code>Spatial_Long_Read_analysis.Rmd</code></div>

    </div>

    
    
<p>We will load the following packages</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">spotglm</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span></code></pre>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this vignette we will go through a typical analysis using spotGLM
on a spatial long-read dataset. The dataset comes from a mouse <a href="https://academic.oup.com/nar/article/51/8/e47/7079641" class="external-link">olfactory
bulb</a>. In the dataset, we identify 11 major cell types;astrocytes,
endothelial cells, mesenchymal cells, microglia, monocytes, mural cells,
oligodendrocytes, neurons (N), olfactory ensheathing cells (OEC),
oligodendrocyte progenitor cells (OPC), and red blood cells. We also are
given 5 distinct brain regions; the olfactory nerve layer, glomerular
layer, outer plexiform layer, mitral layer, and granule layer. We will
identify genes that show evidence of isoform switching in a cell type
specific manner.</p>
<div class="section level3">
<h3 id="read-in-data">Read in Data<a class="anchor" aria-label="anchor" href="#read-in-data"></a>
</h3>
We will read in data with the <code><a href="../reference/read_spatial_long_read.html">read_spatial_long_read()</a></code>
function. The function returns the following
<details><summary>
Output
</summary><ul>
<li>
<code>coords</code>: Coordinate matrix. Of dimension 918 by 2</li>
<li>
<code>niche</code>: The one hot encoded region matrix for each spot.
Of dimension 918 by 5</li>
<li>
<code>deconv</code>: The deconvolution of each spot. of dimension
918 by 11</li>
<li>
<code>library size</code>: The library size of each spot. A vector
of length 918</li>
<li>
<code>total_gene_expression</code>: For each gene, a vector of spot
gene expression. A matrix of dimension 918 by 643</li>
<li>
<code>isoform expression</code>: A list with each element
corresponding to a gene. <code>isoform expression[[gene]]</code> is a
list of length two that contains the spot level expression of the top
two isoforms for that gene.</li>
</ul></details><pre><code><span><span class="va">data</span> <span class="op">=</span> <span class="fu">spotglm</span><span class="fu">::</span><span class="fu"><a href="../reference/read_spatial_long_read.html">read_spatial_long_read</a></span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="plot-regions">Plot Regions<a class="anchor" aria-label="anchor" href="#plot-regions"></a>
</h3>
<pre><code><span></span>
<span><span class="va">region</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">regions</span>,<span class="fl">1</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">regions</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">x</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span><span class="op">]</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  region <span class="op">=</span> <span class="va">region</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">region</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Spots by Region"</span>, color <span class="op">=</span> <span class="st">"Region"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="model-overview">Model Overview<a class="anchor" aria-label="anchor" href="#model-overview"></a>
</h3>
<p>We return to the mouse olfactory bulb data. We restrict our analysis
to only those genes with exactly two isoforms whose summed expression
comprise at least 20% of all isoforms expressed for that gene. We
further limited to genes with and total expression across all spots &gt;
50 UMIs. In total, 643 genes passed these two criteria. For each gene,
we identified cell types where the gene is adequately expressed. A gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
is defined as being adequately expressed within a cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
if the total proportion of transcripts belonging to
gene<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
in cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
is greater than 1/60000. The exact proportion was found by fitting
SpotGLM with only the intercept term. The exact model is</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Y</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi></mrow></msub><mo>∼</mo><mi>P</mi><mi>o</mi><mi>i</mi><mi>s</mi><mrow><mo stretchy="true" form="prefix">(</mo><mo>exp</mo><mrow><mo stretchy="true" form="prefix">(</mo><mo>log</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>L</mi><mi>s</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><munder><mo>∑</mo><mi>t</mi></munder><msub><mi>π</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><mo>exp</mo><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>β</mi><mrow><mn>0</mn><mo>,</mo><mi>g</mi></mrow><mi>t</mi></msubsup><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">Y_{s,g} \sim Pois(\exp(\log(L_s) + \sum_{t} \pi_{s,t}\exp(\beta_{0,g}^t)))</annotation></semantics></math>
In this model,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>L</mi><mi>s</mi></msub><annotation encoding="application/x-tex">L_s</annotation></semantics></math>
represents the library size of spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>,<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>π</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">\pi_{s,t}</annotation></semantics></math>
represents the deconvolution estimate of cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
in spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>,
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>β</mi><mrow><mn>0</mn><mo>,</mo><mi>g</mi></mrow><mi>t</mi></msubsup><annotation encoding="application/x-tex">\beta_{0,g}^t</annotation></semantics></math>
represents the intercept estimate for cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>.
We then focused on these cell types and applied SpotGLM to compare cell
type-specific isoform relative expression between all pairs of
regions.<br></p>
<p>The SpotGLM and corresponding test are described as follows:<br>
For a gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
let the isoform relative proportion the isoform of interest in spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>
be
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>γ</mi><mi>s</mi><mi>g</mi></msubsup><annotation encoding="application/x-tex">\gamma_s^g</annotation></semantics></math>.
The isoform relative proportion is defined as
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>ℙ</mi><mrow><mo stretchy="true" form="prefix">[</mo><mtext mathvariant="normal">UMI is of isoform i |UMI is of gene that isoform belongs to</mtext><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbb{P}[\text{UMI is of isoform i |UMI is of gene that isoform belongs to}]</annotation></semantics></math></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>γ</mi><mi>s</mi><mi>g</mi></msubsup><annotation encoding="application/x-tex">\gamma_s^g</annotation></semantics></math>
is a weighted average of the isoform relative proportions of the cell
types that make up
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>
with cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>’s
weight equal to the cell type mean expression of gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>μ</mi><mrow><mi>g</mi><mo>,</mo><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">\mu_{g,t}</annotation></semantics></math>
times the proportion of cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
in spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>π</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">\pi_{s,t}</annotation></semantics></math>.Let
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>γ</mi><mi>t</mi><mi>g</mi></msubsup><annotation encoding="application/x-tex">\gamma_t^g</annotation></semantics></math>
be the isoform relative proportion for cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>.
Then</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>γ</mi><mi>s</mi><mi>g</mi></msubsup><mo>=</mo><mfrac><mrow><munder><mo>∑</mo><mi>t</mi></munder><msub><mi>π</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><msubsup><mi>γ</mi><mi>t</mi><mi>g</mi></msubsup><msub><mi>μ</mi><mrow><mi>g</mi><mo>,</mo><mi>t</mi></mrow></msub></mrow><mrow><munder><mo>∑</mo><mrow><mi>t</mi><mi>′</mi><mo>∈</mo><mi>s</mi></mrow></munder><msub><mi>π</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi><mi>′</mi></mrow></msub><msub><mi>μ</mi><mrow><mi>g</mi><mo>,</mo><mi>t</mi><mi>′</mi></mrow></msub></mrow></mfrac></mrow><annotation encoding="application/x-tex">\gamma_s^g = \frac{\sum_{t}\pi_{s,t}\gamma_t^g\mu_{g,t}}{\sum_{t'\in s}\pi_{s,t'}\mu_{g,t'}}</annotation></semantics></math>
At the single cell level, we fit a cell type specific logistic
regression model with our covariate matrix 𝑋 being the one hot encoded
region for each cell. That is that
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>Y</mi><mi>i</mi><mi>g</mi></msubsup><mo>∼</mo><mi>B</mi><mi>i</mi><mi>n</mi><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>n</mi><mi>i</mi><mi>g</mi></msubsup><mo>,</mo><msubsup><mi>γ</mi><mi>i</mi><mi>g</mi></msubsup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">Y_i^g \sim Bin(n_{i}^g, \gamma^g_i)</annotation></semantics></math><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mtext mathvariant="normal">logit</mtext><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>γ</mi><mi>i</mi><mi>g</mi></msubsup><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msub><mi>X</mi><mi>i</mi></msub><msup><mi>β</mi><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo><mi>g</mi></mrow></msup></mrow><annotation encoding="application/x-tex">\text{logit}(\gamma^g_i) = X_i\beta^{CT(i),g}</annotation></semantics></math></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">CT(i)</annotation></semantics></math>
represents the cell type of cell
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>Y</mi><mi>i</mi><mi>g</mi></msubsup><annotation encoding="application/x-tex">Y_i^g</annotation></semantics></math>
represents the number of transcripts belong to the isoform of interest.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>n</mi><mi>i</mi><mi>g</mi></msubsup><annotation encoding="application/x-tex">n_{i}^g</annotation></semantics></math>
represents the total number of transcripts belonging to gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>β</mi><mi>r</mi><mrow><mi>t</mi><mo>,</mo><mi>g</mi></mrow></msubsup><annotation encoding="application/x-tex">\beta^{t,g}_r</annotation></semantics></math>
represents the log odds of the event that a transcript belonging to gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
belongs to the isoform of interest in cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
in region
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>r</mi><annotation encoding="application/x-tex">r</annotation></semantics></math>.<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>β</mi><mi>r</mi><mrow><mi>t</mi><mo>,</mo><mi>g</mi></mrow></msubsup><annotation encoding="application/x-tex">\beta^{t,g}_r</annotation></semantics></math>
can be estimated in a cell type specific manner through spotGLM. We
simply have to weight our deconvolution matrix
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>π</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">\pi_{s,t}</annotation></semantics></math>
by the corresponding gene weight
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>μ</mi><mrow><mi>g</mi><mo>,</mo><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">\mu_{g,t}</annotation></semantics></math>.
To test for cell type specific isoform switching between regions
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>r</mi><annotation encoding="application/x-tex">r</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>r</mi><mi>′</mi></mrow><annotation encoding="application/x-tex">r'</annotation></semantics></math>
in cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>,
we test the null hypothesis</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mn>0</mn></msub><mo>:</mo><msubsup><mi>β</mi><mi>r</mi><mrow><mi>t</mi><mo>,</mo><mi>g</mi></mrow></msubsup><mo>=</mo><msubsup><mi>β</mi><mrow><mi>r</mi><mi>′</mi></mrow><mrow><mi>t</mi><mo>,</mo><mi>g</mi></mrow></msubsup></mrow><annotation encoding="application/x-tex">H_0:\beta^{t,g}_r = \beta^{t,g}_{r'}</annotation></semantics></math>
Note that both models require use of the spotglm function
<code>run_model</code>.</p>
<div class="section level4">
<h4 id="step-1-get-average-expression-of-gene-for-each-cell-type-1-minute">Step 1: Get Average Expression of Gene For Each Cell Type (~1
minute)<a class="anchor" aria-label="anchor" href="#step-1-get-average-expression-of-gene-for-each-cell-type-1-minute"></a>
</h4>
<p>We first fit the intercept model as described above.We make use of
the spotglm function “run_model”. We specify our response
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>y</mi><annotation encoding="application/x-tex">y</annotation></semantics></math>
which we take to be the gene expression values for a gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>.
our covariates
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
is just the intercept matrix.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>λ</mi><annotation encoding="application/x-tex">\lambda</annotation></semantics></math>
is our deconvolution which we make sure has row sums that sum to 1. Our
offset is the library size of the each spot. The initialization used is
“full” since we want to initialize all parameters. Our family is “spot
poisson” since we are fitting a poisson model.</p>
<pre><code><span><span class="co">#get number of genes </span></span>
<span><span class="va">ngenes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">total_gene_expression</span><span class="op">)</span></span>
<span><span class="co">#initialize intercept estimates</span></span>
<span><span class="va">intercept_estimate</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="st">"list"</span>,<span class="va">ngenes</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">intercept_estimate</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">total_gene_expression</span><span class="op">)</span></span>
<span><span class="co">#initialize list of important cell types for this gene </span></span>
<span><span class="va">good_cell_type_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="st">"list"</span>,<span class="va">ngenes</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">good_cell_type_list</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">total_gene_expression</span><span class="op">)</span></span>
<span><span class="co">#iterate over each gene </span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">ngenes</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">intercept_estimate</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span><span class="op">=</span><span class="fu">spotglm</span><span class="fu">::</span><span class="fu"><a href="../reference/run_model.html">run_model</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">total_gene_expression</span><span class="op">[</span>,<span class="va">j</span><span class="op">]</span>,</span>
<span>                               X <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">total_gene_expression</span><span class="op">)</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>                               lambda <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">deconv</span>,</span>
<span>                               offset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">library_size</span><span class="op">)</span>,</span>
<span>                               initialization <span class="op">=</span> <span class="st">"full"</span>,</span>
<span>                               family <span class="op">=</span> <span class="st">"spot poisson"</span>,intercept_col <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                               min_reads_per_1000 <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">#get T statistics for intercepts</span></span>
<span>  <span class="va">T_statistics</span> <span class="op">=</span> <span class="va">intercept_estimate</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">beta_est</span></span>
<span>  <span class="va">T_statistics</span> <span class="op">=</span> <span class="va">T_statistics</span><span class="op">/</span><span class="va">intercept_estimate</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">stand_err_mat</span></span>
<span>  </span>
<span>  <span class="co">#reweight deconvolution vector to reflect difference in expression </span></span>
<span>  <span class="va">deconv_weighted</span> <span class="op">=</span> <span class="fl">0</span><span class="op">*</span><span class="va">data</span><span class="op">$</span><span class="va">deconv</span></span>
<span>  <span class="kw">for</span><span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">deconv_weighted</span><span class="op">[</span><span class="va">k</span>,<span class="op">]</span><span class="op">=</span><span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">[</span><span class="va">k</span>,<span class="op">]</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">intercept_estimate</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">beta_est</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co">#normalize to sum to one</span></span>
<span>  <span class="va">deconv_weighted</span> <span class="op">=</span> <span class="va">deconv_weighted</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">deconv_weighted</span><span class="op">)</span></span>
<span>  <span class="co">#remove those cells that have small weights</span></span>
<span>  <span class="va">deconv_weighted</span><span class="op">[</span><span class="va">deconv_weighted</span> <span class="op">&lt;</span> <span class="fl">0.2</span><span class="op">]</span> <span class="op">=</span> <span class="fl">0</span></span>
<span>  <span class="co">#re-normalize</span></span>
<span>  <span class="va">deconv_weighted</span> <span class="op">=</span> <span class="va">deconv_weighted</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">deconv_weighted</span><span class="op">)</span></span>
<span>  <span class="co">#get cell type frequencies</span></span>
<span>  <span class="va">ct_freq</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">deconv_weighted</span>,<span class="fl">2</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span> <span class="op">&gt;</span> <span class="fl">0</span>,na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span>  <span class="co">#remove cell types that don't appear more than 50 times</span></span>
<span>  <span class="va">important_ct</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">ct_freq</span> <span class="op">&gt;</span> <span class="fl">50</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">#get "good" cell types: Those that have sufficient expression of the gene </span></span>
<span>  <span class="va">good_ct</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">T_statistics</span> <span class="op">&lt;</span> <span class="op">-</span><span class="fl">1.96</span> <span class="op">&amp;</span> <span class="va">intercept_estimate</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">beta_est</span> <span class="op">&gt;</span> <span class="op">-</span><span class="fl">11</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">#get intersection,i.e cell types with sufficient expression and sample size</span></span>
<span>  <span class="va">good_cell_type_list</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">good_ct</span>, <span class="va">important_ct</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="step-2-spotglm-to-perform-isoform-differential-expression-1-minute">Step 2: SpotGLM to Perform Isoform Differential Expression (~1
minute)<a class="anchor" aria-label="anchor" href="#step-2-spotglm-to-perform-isoform-differential-expression-1-minute"></a>
</h4>
<p>We next fit the binomial spotGLM as described above. We again use the
“run model” function. This time however,the family is “spot binomial”
since we are running a binomial model. We additionally use two
parameters that are generally not used when running spotglm
<code>weights</code> and <code>ct_cov_weights</code>. In a binomial
regression model, it is necessary to weight observations by the number
of trials used. For each gene this corresponds to the total gene
expression vector. Additionally, unlike the spot poisson model where the
spot expression was a weighted mean of cell type level expression with
the weight corresponding to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>π</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">\pi_{s,t}</annotation></semantics></math>,
in the spot binomial model, there is an additional cell type specific
weight
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>μ</mi><mrow><mi>g</mi><mo>,</mo><mi>t</mi></mrow></msub><mo>=</mo><mo>exp</mo><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>β</mi><mrow><mn>0</mn><mo>,</mo><mi>g</mi></mrow><mi>t</mi></msubsup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\mu_{g,t} = \exp(\beta_{0,g}^t)</annotation></semantics></math>.
We can include this vector in the <code>ct_cov_weights</code> argument.
We fit the binomial model below.</p>
<pre><code><span></span>
<span><span class="co">#Step 2: Fit spot binomial model</span></span>
<span><span class="va">isoform_DE</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="st">"list"</span>,<span class="va">ngenes</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">isoform_DE</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">total_gene_expression</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">ngenes</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co">#get gene name </span></span>
<span>  <span class="va">gene</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">isoform_DE</span><span class="op">)</span><span class="op">[</span><span class="va">j</span><span class="op">]</span></span>
<span>  <span class="co">#get weights </span></span>
<span>  <span class="va">weights</span> <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">total_gene_expression</span><span class="op">[</span>,<span class="va">gene</span><span class="op">]</span></span>
<span>  <span class="co">#get covariate weights </span></span>
<span>  <span class="va">ct_covariate_weights</span>  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">intercept_estimate</span><span class="op">[[</span><span class="va">gene</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">beta_est</span><span class="op">)</span></span>
<span>  <span class="co">#get good cell types for this gene</span></span>
<span>  <span class="va">good_cell_types</span> <span class="op">=</span> <span class="va">good_cell_type_list</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">good_cell_types</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw">next</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="co">#run model </span></span>
<span>  <span class="va">isoform_DE</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu">spotglm</span><span class="fu">::</span><span class="fu"><a href="../reference/run_model.html">run_model</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">isoform_expression</span><span class="op">[[</span><span class="va">gene</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                         X <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">regions</span>,</span>
<span>                         lambda <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">[</span>,<span class="va">good_cell_types</span>,drop <span class="op">=</span> <span class="cn">F</span><span class="op">]</span>,</span>
<span>                         family <span class="op">=</span> <span class="st">"spot binomial"</span>,</span>
<span>                         weights <span class="op">=</span> <span class="va">weights</span>,</span>
<span>                         ct_cov_weights <span class="op">=</span> <span class="va">ct_covariate_weights</span><span class="op">[</span><span class="va">good_cell_types</span><span class="op">]</span>,</span>
<span>                         initialization <span class="op">=</span> <span class="st">"full"</span>,</span>
<span>                         min_freq <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="step-3-downstream-analysis-contrast-test-to-identify-isoform-switching-across-regions">Step 3: Downstream Analysis: Contrast Test to Identify Isoform
Switching Across Regions<a class="anchor" aria-label="anchor" href="#step-3-downstream-analysis-contrast-test-to-identify-isoform-switching-across-regions"></a>
</h4>
<p>We now test for cell type specific isoform switching across regions.
This uses the <code>compute_contrast_test()</code> function. We append
the matrices outputted from the “compute contrast significance” function
and perform a global BH correction. We then print out the genes that
show significant isoform switching.</p>
<pre><code><span><span class="co">#Store all contrast tests </span></span>
<span><span class="va">all_contrast_tests</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">cell_types</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">)</span></span>
<span><span class="va">regions</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">regions</span><span class="op">)</span></span>
<span><span class="va">nregion</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">regions</span><span class="op">)</span></span>
<span><span class="va">counter</span> <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="co">#test over all cell types and region pairs </span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">ct</span> <span class="kw">in</span> <span class="va">cell_types</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">nregion</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw">for</span><span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">j</span><span class="op">:</span><span class="va">nregion</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>      <span class="kw">if</span><span class="op">(</span><span class="va">j</span><span class="op">==</span><span class="va">k</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="kw">next</span></span>
<span>      <span class="op">}</span></span>
<span>      <span class="va">region_A</span> <span class="op">=</span> <span class="va">regions</span><span class="op">[</span><span class="va">j</span><span class="op">]</span></span>
<span>      <span class="va">region_B</span> <span class="op">=</span> <span class="va">regions</span><span class="op">[</span><span class="va">k</span><span class="op">]</span></span>
<span>      <span class="co">#compute pvalues</span></span>
<span>      <span class="va">pvals</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu">spotglm</span><span class="fu">::</span><span class="fu"><a href="../reference/compute_contrast_significance.html">compute_contrast_significance</a></span><span class="op">(</span>input_list <span class="op">=</span> <span class="va">isoform_DE</span>,</span>
<span>                                           cell_type <span class="op">=</span> <span class="va">ct</span>,</span>
<span>                                           effect_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">region_A</span>,<span class="va">region_B</span><span class="op">)</span>,</span>
<span>                                           beta_name <span class="op">=</span> <span class="st">"beta_est"</span>,</span>
<span>                                           covariance_name <span class="op">=</span> <span class="st">"vcov"</span>,</span>
<span>                                           sided <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>      </span>
<span>      <span class="va">contrast_tests</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">pvals</span><span class="op">$</span><span class="va">name</span>, cell_type <span class="op">=</span> <span class="va">ct</span>, </span>
<span>                                  region_A <span class="op">=</span> <span class="va">region_A</span>, region_B <span class="op">=</span> <span class="va">region_B</span>, </span>
<span>                                  pval <span class="op">=</span> <span class="va">pvals</span><span class="op">$</span><span class="va">pval</span><span class="op">)</span></span>
<span>      <span class="co">#add results to list </span></span>
<span>      <span class="va">all_contrast_tests</span><span class="op">[[</span><span class="va">counter</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">contrast_tests</span></span>
<span>      <span class="va">counter</span> <span class="op">=</span> <span class="va">counter</span> <span class="op">+</span> <span class="fl">1</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span><span class="va">combined_contrast_tests</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">all_contrast_tests</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                                                    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">pval</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#get significant switches</span></span>
<span><span class="va">combined_contrast_tests</span><span class="op">$</span><span class="va">qval</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">combined_contrast_tests</span><span class="op">$</span><span class="va">pval</span>,method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span><span class="va">significant_isoform_switches</span> <span class="op">=</span> <span class="va">combined_contrast_tests</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">qval</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span></span></code></pre>
</div>
<div class="section level4">
<h4 id="results-of-contrast-test-as-a-table">Results of Contrast Test as a Table<a class="anchor" aria-label="anchor" href="#results-of-contrast-test-as-a-table"></a>
</h4>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">significant_isoform_switches</span>,<span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="example-gene-ap3s1">Example Gene: Ap3s1<a class="anchor" aria-label="anchor" href="#example-gene-ap3s1"></a>
</h4>
<p>Here we show a gene, Ap3s1, that shows evidence of isoform switching
between the granule layer and other layers in Neurons. Because Neurons
are the only important cell types for this gene, we can safely just plot
overall isoform relative proportion and attribute any differences to
Neurons.</p>
<pre><code><span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="st">"Ap3s1"</span></span>
<span></span>
<span><span class="co"># Isoform ratio</span></span>
<span><span class="va">isoform_vals</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">isoform_expression</span><span class="op">[[</span><span class="va">gene</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">gene_total</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">total_gene_expression</span><span class="op">[</span>, <span class="va">gene</span><span class="op">]</span></span>
<span><span class="va">ratio</span> <span class="op">&lt;-</span> <span class="va">isoform_vals</span> <span class="op">/</span> <span class="va">gene_total</span></span>
<span><span class="va">ratio</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/is.finite.html" class="external-link">is.finite</a></span><span class="op">(</span><span class="va">ratio</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span></span>
<span><span class="co"># Region labels</span></span>
<span><span class="va">region</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">regions</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">regions</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">x</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Binary group</span></span>
<span><span class="va">region_group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">region</span> <span class="op">==</span> <span class="st">"regionGranule Cell Layer (GCL+RMS)"</span>, <span class="st">"Granule Layer"</span>, <span class="st">"Other"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Combined dataframe</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>,</span>
<span>  ratio <span class="op">=</span> <span class="va">ratio</span>,</span>
<span>  region <span class="op">=</span> <span class="va">region_group</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">ratio</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># === Plot 1: Violin + boxplot ===</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">region</span>, y <span class="op">=</span> <span class="va">ratio</span>, fill <span class="op">=</span> <span class="va">region</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html" class="external-link">geom_violin</a></span><span class="op">(</span>trim <span class="op">=</span> <span class="cn">FALSE</span>, alpha <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.2</span>, outlier.shape <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">gene</span>, <span class="st">"Isoform/Gene Ratio"</span><span class="op">)</span>,</span>
<span>    x <span class="op">=</span> <span class="st">""</span>, y <span class="op">=</span> <span class="st">"Ratio"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Granule Layer"</span> <span class="op">=</span> <span class="st">"tomato"</span>, <span class="st">"Other"</span> <span class="op">=</span> <span class="st">"skyblue"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># === Plot 2: Spatial plot - Granule Layer only ===</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="va">df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">region</span> <span class="op">==</span> <span class="st">"Granule Layer"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">ratio</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span>na.value <span class="op">=</span> <span class="st">"grey90"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Granule Layer"</span>, x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span>, color <span class="op">=</span> <span class="st">"Ratio"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># === Plot 3: Spatial plot - Other regions ===</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="va">df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">region</span> <span class="op">==</span> <span class="st">"Other"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">ratio</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span>na.value <span class="op">=</span> <span class="st">"grey90"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Other Regions"</span>, x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span>, color <span class="op">=</span> <span class="st">"Ratio"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">p1</span></span>
<span><span class="va">p2</span> <span class="op">+</span> <span class="va">p3</span></span>
<span></span>
<span></span></code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kaishu Mason.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
