<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Spatial ATAC Analysis ‚Ä¢ spotGLM</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Spatial ATAC Analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">spotGLM</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Intro_to_SpotGLM.html">Intro to SpotGLM</a></li>
    <li><a class="dropdown-item" href="../articles/Spatial_ATAC_analysis.html">Spatial ATAC Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Spatial_Long_Read_analysis.html">Spatial Long-Read Isoform Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Vignette_VisiumHD_Mouse_Kidney_analysis.html">Visium HD Workflow - SpotGLM w/ SPARROW</a></li>
    <li><a class="dropdown-item" href="../articles/Visium_analysis.html">Visium Niche DE Analysis</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/kaishumason/spotGLM/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Spatial ATAC Analysis</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/kaishumason/spotGLM/blob/HEAD/vignettes/Spatial_ATAC_analysis.Rmd"><code>vignettes/Spatial_ATAC_analysis.Rmd</code></a></small>
      <div class="d-none name"><code>Spatial_ATAC_analysis.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this vignette we will go through a typical analysis using spotGLM
on a spatial ATAC dataset. The dataset comes from a <a href="https://www.nature.com/articles/s41586-023-05795-1#Fig1" class="external-link">p22 mouse
brain from this spatial epigenome-transcriptome co-profiling study by Di
Zhang et al.¬†2023</a>. In the dataset, there are 13 major cell types,
among which we will focus on oligodendrocytes. The study also identified
14 distinct regions, among which we will focus on comparing between the
cortex (C0) and the corpus callosum (C3). We have called peaks and used
ChromVAR to compute transcription factor activity scores by aggregating
reads that overlap with their motif signatures. In this tutorial, our
focus will be to identify changes in transcription factor motif
accessibility across the two brain regions in a cell type specific
manner.</p>
</div>
<div class="section level2">
<h2 id="setup-and-data-loading">Setup and data loading<a class="anchor" aria-label="anchor" href="#setup-and-data-loading"></a>
</h2>
<p>We start by loading the following packages.</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kaishumason/spotGLM">spotGLM</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span> <span class="co"># for plotting</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span> <span class="co"># for finetuning the appearance of ggplots </span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span> <span class="co"># for manipulating data tables</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span> <span class="co"># for qqplots side-by-side</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggrepel.slowkow.com/" class="external-link">ggrepel</a></span><span class="op">)</span> <span class="co"># for labeling points in qqplot</span></span></code></pre>
<p>The processed brain spatial ATAC data is stored in our public github
repository. We will read in the data with the
<code>read_spatial_mouse_brain_atac_data()</code> function. The function
returns the following</p>
<ul>
<li>
<code>coords</code>: Coordinate matrix. Of dimension 9215 by 2</li>
<li>
<code>regions</code>: The one hot encoded region matrix for each
spot. Of dimension 9215 by 14</li>
<li>
<code>deconv</code>: The deconvolution of each spot. of dimension
9215 by 9</li>
<li>
<code>motif_scores</code>: The motif score for each spot. A matrix
of dimension 9215 by 746.</li>
</ul>
<pre><code><span><span class="va">data</span> <span class="op">=</span> <span class="fu">spotGLM</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/spotGLM/man/read_example_spatial_mouse_brain_atac_data.html" class="external-link">read_example_spatial_mouse_brain_atac_data</a></span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="plot-of-spatial-regions">Plot of Spatial Regions<a class="anchor" aria-label="anchor" href="#plot-of-spatial-regions"></a>
</h2>
<p>First, let‚Äôs get a feel for this data by plotting a map of the
spatial regions.</p>
<pre><code><span><span class="co">#number of regions</span></span>
<span><span class="va">region</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">regions</span>,<span class="fl">1</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">x</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">n_regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">region</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a data frame for plotting</span></span>
<span><span class="va">plot_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  region <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">region</span><span class="op">)</span>  <span class="co"># convert to factor for discrete coloring</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span><span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">region</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://scales.r-lib.org/reference/pal_hue.html" class="external-link">hue_pal</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="va">n_regions</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Spots Colored by Region"</span>, color <span class="op">=</span> <span class="st">"Region"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"brain_atac_region_map.png"</span>, plot<span class="op">=</span><span class="va">p1</span>, width<span class="op">=</span><span class="fl">8</span>, height<span class="op">=</span><span class="fl">8</span>, dpi<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span></code></pre>
<p><img src="../reference/figures/brain_atac_region_map.png" alt="Brain ATAC spatial region map" style="width:80%;"></p>
</div>
<div class="section level2">
<h2 id="distribution-of-motif-scores">Distribution of motif scores<a class="anchor" aria-label="anchor" href="#distribution-of-motif-scores"></a>
</h2>
<p>Unlike the case of gene expression or isoform expression in the other
tutorials, the distribution of motif scores are approximately normal, as
we can see by randomly picking a few and visualizing their
histograms:</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"brain_atac_motif_score_distributions.png"</span>, height<span class="op">=</span><span class="fl">8</span>, width<span class="op">=</span><span class="fl">8</span>, res<span class="op">=</span><span class="fl">300</span>, units<span class="op">=</span><span class="st">"in"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">motif_scores</span><span class="op">)</span>, <span class="fl">9</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">motif_scores</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>, xlab<span class="op">=</span><span class="st">"Motif Scores"</span>, main<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">motif_scores</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, col<span class="op">=</span><span class="st">"pink"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre>
<p><img src="../reference/figures/brain_atac_motif_score_distributions.png" alt="Brain ATAC motif score distributions" style="width:80%;"></p>
</div>
<div class="section level2">
<h2 id="details-on-statistical-model">Details on statistical model<a class="anchor" aria-label="anchor" href="#details-on-statistical-model"></a>
</h2>
<p>Above, we found that the motif scores have approximately normally
shaped distributions, so we will use a normal error model. First suppose
that we have single cell data. Let
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Y</mi><mrow><mi>i</mi><mo>,</mo><mi>m</mi></mrow></msub><annotation encoding="application/x-tex">Y_{i,m}</annotation></semantics></math>
be the motif score of cell
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
for motif
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>.
We would model
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Y</mi><mrow><mi>i</mi><mo>,</mo><mi>m</mi></mrow></msub><annotation encoding="application/x-tex">Y_{i,m}</annotation></semantics></math>
as follows
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Y</mi><mrow><mi>i</mi><mo>,</mo><mi>m</mi></mrow></msub><mo>‚àº</mo><mi>ùí©</mi><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>Œ≤</mi><mrow><mi>m</mi><mo>,</mo><msub><mi>r</mi><mi>i</mi></msub></mrow><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msubsup><mo>+</mo><msubsup><mi>Œ≤</mi><mrow><mi>m</mi><mo>,</mo><mn>0</mn></mrow><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msubsup><mo>,</mo><msubsup><mi>œÉ</mi><mi>m</mi><mn>2</mn></msubsup><mo stretchy="true" form="postfix">)</mo></mrow><mi>.</mi></mrow><annotation encoding="application/x-tex">Y_{i,m} \sim \mathcal{N}(\beta^{CT(i)}_{m,r_i} + \beta^{CT(i)}_{m,0},\sigma_m^2).</annotation></semantics></math>
Here
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">CT(i)</annotation></semantics></math>
represents the cell type of cell
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>r</mi><mi>i</mi></msub><annotation encoding="application/x-tex">r_i</annotation></semantics></math>
represents the region that cell
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
resides in.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>Œ≤</mi><mrow><mi>m</mi><mo>,</mo><msub><mi>r</mi><mi>i</mi></msub></mrow><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msubsup><annotation encoding="application/x-tex">\beta^{CT(i)}_{m,r_i}</annotation></semantics></math>
represents the expected motif score of a cell of type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">CT(i)</annotation></semantics></math>
in region
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>r</mi><mi>i</mi></msub><annotation encoding="application/x-tex">r_i</annotation></semantics></math>,
while
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>Œ≤</mi><mrow><mi>m</mi><mo>,</mo><mn>0</mn></mrow><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msubsup><annotation encoding="application/x-tex">\beta^{CT(i)}_{m,0}</annotation></semantics></math>
is an intercept term.</p>
<p>For spot-level data, let
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Y</mi><mrow><mi>s</mi><mo>,</mo><mi>m</mi></mrow></msub><annotation encoding="application/x-tex">Y_{s,m}</annotation></semantics></math>
be the score of motif
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>
in spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>.
We assume that this spot-level score is a weighted average of the motif
scores for all cells that reside in it, i.e
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Y</mi><mrow><mi>s</mi><mo>,</mo><mi>m</mi></mrow></msub><mo>‚àº</mo><mi>ùí©</mi><mrow><mo stretchy="true" form="prefix">(</mo><munder><mo>‚àë</mo><mi>t</mi></munder><msub><mi>œÄ</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>Œ≤</mi><mrow><mi>m</mi><mo>,</mo><msub><mi>r</mi><mi>s</mi></msub></mrow><mi>t</mi></msubsup><mo>+</mo><msubsup><mi>Œ≤</mi><mrow><mi>m</mi><mo>,</mo><mn>0</mn></mrow><mi>t</mi></msubsup><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo><msubsup><mi>œÉ</mi><mi>m</mi><mn>2</mn></msubsup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">Y_{s,m} \sim \mathcal{N}( \sum_{t}\pi_{s,t}(\beta^{t}_{m,r_s} + \beta^{t}_{m,0}),\sigma_m^2)</annotation></semantics></math>
where the weight for cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>œÄ</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">\pi_{s,t}</annotation></semantics></math>,
is the deconvolution estimate of the contribution of cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
to spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>.
To test for differences in motif accessibility between regions in a cell
type specific manner, we compute the p-value against the null hypothesis
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mn>0</mn></msub><mo>:</mo><msubsup><mi>Œ≤</mi><mrow><mi>m</mi><mo>,</mo><mi>r</mi></mrow><mi>t</mi></msubsup><mo>=</mo><msubsup><mi>Œ≤</mi><mrow><mi>m</mi><mo>,</mo><mi>r</mi><mi>‚Ä≤</mi></mrow><mi>t</mi></msubsup><mi>.</mi></mrow><annotation encoding="application/x-tex">H_0: \beta^{t}_{m,r} = \beta^{t}_{m,r'}.</annotation></semantics></math></p>
</div>
<div class="section level2">
<h2 id="fitting-spotglm-model-across-all-cell-types-and-regions-2-minutes">Fitting spotGLM model across all cell types and regions (~2
minutes)<a class="anchor" aria-label="anchor" href="#fitting-spotglm-model-across-all-cell-types-and-regions-2-minutes"></a>
</h2>
<p>Since we are using the normal noise model here, we are effectively
fitting a linear model. Thus, we can use the <code>spot_lm</code>
function within spotGLM (<code>lm</code> standards for
<code>linear model</code>). The code below fits the model for each
motif, effectively computing the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>Œ≤</mi><annotation encoding="application/x-tex">\beta</annotation></semantics></math>
parameters for each motif, each cell type, in each region. The results
are stored in <code>res</code>, which can be used downstream to test
hypotheses of interest. This is the time consuming part of the analysis,
taking roughly 2 minutes.</p>
<pre><code><span><span class="va">num_motifs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">motif_scores</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="st">"list"</span>,<span class="va">num_motifs</span><span class="op">)</span></span>
<span></span>
<span><span class="va">t1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">num_motifs</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="va">j</span><span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span><span class="fl">100</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Fitting model for motif "</span>,<span class="va">j</span>,<span class="st">" out of "</span>, <span class="va">num_motifs</span>,<span class="st">"\n"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">t1</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">res</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="../reference/spot_lm.html">spot_lm</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">motif_scores</span><span class="op">[</span>,<span class="va">j</span><span class="op">]</span>,X <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">regions</span>,</span>
<span>                     lambda <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">motif_scores</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="oligodendrocyte-differential-motif-accessibility">Oligodendrocyte Differential Motif Accessibility<a class="anchor" aria-label="anchor" href="#oligodendrocyte-differential-motif-accessibility"></a>
</h2>
<p>We now focus on in oligodendrocytes in the corpus callosum vs cortex
region and use the <code><a href="../reference/compute_contrast_significance.html">compute_contrast_significance()</a></code> function
to test the hypothesis, for each motif, whether its accessibility
differs in oligodendrocytes between the two regions.</p>
<pre><code><span><span class="va">sig_genes</span> <span class="op">=</span> <span class="fu"><a href="../reference/compute_contrast_significance.html">compute_contrast_significance</a></span><span class="op">(</span>input_list <span class="op">=</span> <span class="va">res</span>,</span>
<span>                                    cell_type <span class="op">=</span> <span class="st">"Oligodendrocytes"</span>,</span>
<span>                                    effect_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"regionC3"</span>,<span class="st">"regionC0"</span><span class="op">)</span>,</span>
<span>                                    beta_name <span class="op">=</span> <span class="st">"beta_estimate"</span>,covariance_name <span class="op">=</span> <span class="st">"vcov"</span>,</span>
<span>                                    sided <span class="op">=</span> <span class="fl">1</span>,direction <span class="op">=</span> <span class="st">"pos"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sig_genes</span> <span class="op">=</span> <span class="va">sig_genes</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">pval</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<div class="section level3">
<h3 id="visualize-test-significance-via-qqplot">Visualize test significance via QQplot<a class="anchor" aria-label="anchor" href="#visualize-test-significance-via-qqplot"></a>
</h3>
<p>The code below draws a qq-plot of the motif test-statistics, labeling
the top motifs. We see that the Sox transcription factors rank at the
top.</p>
<pre><code><span><span class="va">sig_genes_df</span> <span class="op">&lt;-</span> <span class="va">sig_genes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">sig_genes</span><span class="op">$</span><span class="va">test_statistic</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">sig_genes_df</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sig_genes_df</span><span class="op">)</span></span>
<span></span>
<span><span class="va">qq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/qqnorm.html" class="external-link">qqnorm</a></span><span class="op">(</span><span class="va">sig_genes_df</span><span class="op">$</span><span class="va">test_statistic</span>, plot.it <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">sig_genes_df</span><span class="op">$</span><span class="va">theoretical</span> <span class="op">&lt;-</span> <span class="va">qq</span><span class="op">$</span><span class="va">x</span>  <span class="co"># theoretical quantiles</span></span>
<span><span class="va">sig_genes_df</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="va">qq</span><span class="op">$</span><span class="va">y</span>       <span class="co"># sample quantiles</span></span>
<span><span class="co"># Compute the slope and intercept for the QQ line</span></span>
<span><span class="va">q1_sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">sig_genes_df</span><span class="op">$</span><span class="va">test_statistic</span>, <span class="fl">0.25</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">q3_sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">sig_genes_df</span><span class="op">$</span><span class="va">test_statistic</span>, <span class="fl">0.75</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">q1_theor</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">qnorm</a></span><span class="op">(</span><span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="va">q3_theor</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">qnorm</a></span><span class="op">(</span><span class="fl">0.75</span><span class="op">)</span></span>
<span><span class="va">slope</span>     <span class="op">&lt;-</span> <span class="op">(</span><span class="va">q3_sample</span> <span class="op">-</span> <span class="va">q1_sample</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">q3_theor</span> <span class="op">-</span> <span class="va">q1_theor</span><span class="op">)</span></span>
<span><span class="va">intercept</span> <span class="op">&lt;-</span> <span class="va">q1_sample</span> <span class="op">-</span> <span class="va">slope</span> <span class="op">*</span> <span class="va">q1_theor</span></span>
<span></span>
<span><span class="co"># Select the top genes by highest t_test_stat</span></span>
<span><span class="va">top_n</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="va">top_df</span> <span class="op">&lt;-</span> <span class="va">sig_genes_df</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">sig_genes_df</span><span class="op">$</span><span class="va">test_statistic</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">top_n</span><span class="op">]</span>, <span class="op">]</span></span>
<span><span class="co"># Create the QQ plot using ggplot2 with ggrepel for labeled boxes</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">sig_genes_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">theoretical</span>, y <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>intercept <span class="op">=</span> <span class="va">intercept</span>, slope <span class="op">=</span> <span class="va">slope</span>, color <span class="op">=</span> <span class="st">"red"</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html" class="external-link">geom_label_repel</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">top_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">gene</span><span class="op">)</span>,</span>
<span>                   box.padding <span class="op">=</span> <span class="fl">0.35</span>, point.padding <span class="op">=</span> <span class="fl">0.5</span>, <span class="co"># size = 10, </span></span>
<span>                   segment.color <span class="op">=</span> <span class="st">"grey50"</span>, max.overlaps <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Theoretical Quantiles"</span>, </span>
<span>       y <span class="op">=</span> <span class="st">"Sample Quantiles"</span>, </span>
<span>       title <span class="op">=</span> <span class="st">"Oligodendrocytes: Cortex vs Corpus Callosum"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span>, hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"brain_atac_qqplot.png"</span>, plot<span class="op">=</span><span class="va">p</span>, width<span class="op">=</span><span class="fl">8</span>, height<span class="op">=</span><span class="fl">8</span>, dpi<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span></code></pre>
<p><img src="../reference/figures/brain_atac_ggplot.png" alt="Brain ATAC spatial region map" style="width:80%;"></p>
</div>
</div>
<div class="section level2">
<h2 id="example-motif-sox4">Example Motif: SOX4<a class="anchor" aria-label="anchor" href="#example-motif-sox4"></a>
</h2>
<p>We now focus on one example motif, Sox4, that shows differential
motif accessibility in oligodendrocytes between the corpus callosum and
the cortex. Let‚Äôs visualize the distribution of this motif in space, and
compare it to the distribution of oligodendrocytes.</p>
<pre><code><span><span class="va">motif</span> <span class="op">&lt;-</span> <span class="st">"SOX4"</span></span></code></pre>
<p>To make the visual contrast easier to see, we only plot the spots
lying in regionC3 (Corpus Callosum) and regionC0 (Cortex):</p>
<pre><code><span><span class="va">region_filter</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">regions</span><span class="op">[</span>, <span class="st">"regionC3"</span><span class="op">]</span> <span class="op">==</span> <span class="fl">1</span> <span class="op">|</span> <span class="va">data</span><span class="op">$</span><span class="va">regions</span><span class="op">[</span>, <span class="st">"regionC0"</span><span class="op">]</span> <span class="op">==</span> <span class="fl">1</span></span>
<span><span class="co"># Region labels</span></span>
<span><span class="va">region_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">regions</span><span class="op">[</span><span class="va">region_filter</span>, <span class="st">"regionC3"</span><span class="op">]</span> <span class="op">==</span> <span class="fl">1</span>, <span class="st">"Corpus Callosum"</span>, <span class="st">"Cortex"</span><span class="op">)</span></span></code></pre>
<p>We need to truncate the motif scores so that outliers don‚Äôt drive the
color axis:</p>
<pre><code><span><span class="co"># Motif score with truncation</span></span>
<span><span class="va">Sox_score</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">motif_scores</span><span class="op">[</span><span class="va">region_filter</span>, <span class="va">motif</span><span class="op">]</span></span>
<span><span class="va">Sox_score</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmin</a></span><span class="op">(</span><span class="va">Sox_score</span>, <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">Sox_score</span>, <span class="fl">0.99</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">Sox_score</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmax</a></span><span class="op">(</span><span class="va">Sox_score</span>, <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">Sox_score</span>, <span class="fl">0.01</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<pre><code><span><span class="co"># Oligodendrocyte proportion</span></span>
<span><span class="va">oligo_score</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">[</span><span class="va">region_filter</span>, <span class="st">"Oligodendrocytes"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Create plotting data frame</span></span>
<span><span class="va">plot_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span><span class="va">region_filter</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span><span class="va">region_filter</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  Sox <span class="op">=</span> <span class="va">Sox_score</span>,</span>
<span>  region <span class="op">=</span> <span class="va">region_label</span>,</span>
<span>  oligodendrocytes <span class="op">=</span> <span class="va">oligo_score</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># === Plot 1: SOX10 Motif Scores ===</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">Sox</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">motif</span>, <span class="st">" Motif Scores"</span><span class="op">)</span>, color <span class="op">=</span> <span class="va">motif</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"brain_atac_sox4_motif_map.png"</span>, plot<span class="op">=</span><span class="va">p1</span>, width<span class="op">=</span><span class="fl">8</span>, height<span class="op">=</span><span class="fl">8</span>, dpi<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span></code></pre>
<p><img src="../reference/figures/brain_atac_sox4_motif_map.png" alt="Brain ATAC Sox4 motif map" style="width:80%;"></p>
<pre><code><span><span class="co"># === Plot 2: Region membership ===</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">region</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cortex"</span> <span class="op">=</span> <span class="st">"#1f78b4"</span>, <span class="st">"Corpus Callosum"</span> <span class="op">=</span> <span class="st">"#33a02c"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Region Membership"</span>, color <span class="op">=</span> <span class="st">"Region"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"brain_atac_cortex_corpuscallosum_map.png"</span>, plot<span class="op">=</span><span class="va">p2</span>, width<span class="op">=</span><span class="fl">8</span>, height<span class="op">=</span><span class="fl">8</span>, dpi<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span></code></pre>
<p><img src="../reference/figures/brain_atac_cortex_corpuscallosum_map.png" alt="Brain ATAC cortex versus corpuscallosum map" style="width:80%;"></p>
<pre><code><span><span class="co"># === Plot 3: Oligodendrocyte proportion ===</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">oligodendrocytes</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Oligodendrocyte Deconvolution Proportion"</span>, color <span class="op">=</span> <span class="st">"Oligos"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"brain_atac_oligodendrocyte_map.png"</span>, plot<span class="op">=</span><span class="va">p3</span>, width<span class="op">=</span><span class="fl">8</span>, height<span class="op">=</span><span class="fl">8</span>, dpi<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span></code></pre>
<p><img src="../reference/figures/brain_atac_oligodendrocyte_map.png" alt="Brain ATAC oligodendrocyte map" style="width:80%;"></p>
<p>We see that oligodendrocytes are present at high concentrations in
corpus callosum, but are also present throughout the cortex. Through the
spotGLM estimation, we are able to attribute the increase in Sox4
expression in the corpus callosum to the increased expression of this
gene in the oligodendrocytes in the region.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kaishu Mason, Yijia Jiang, Nancy R. Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
