<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Visium HD Workflow - SpotGLM w/ SPARROW • spotGLM</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Visium HD Workflow - SpotGLM w/ SPARROW">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">spotGLM</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Intro_to_SpotGLM.html">Intro to SpotGLM</a></li>
    <li><a class="dropdown-item" href="../articles/Spatial_ATAC_analysis.html">Spatial ATAC Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Spatial_Long_Read_analysis.html">Spatial Long-Read Isoform Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Vignette_VisiumHD_Mouse_Kidney_analysis.html">Visium HD Workflow - SpotGLM w/ SPARROW</a></li>
    <li><a class="dropdown-item" href="../articles/Visium_analysis.html">Visium Niche DE Analysis</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/kaishumason/spotGLM/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Visium HD Workflow - SpotGLM w/ SPARROW</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/kaishumason/spotGLM/blob/HEAD/vignettes/Vignette_VisiumHD_Mouse_Kidney_analysis.Rmd"><code>vignettes/Vignette_VisiumHD_Mouse_Kidney_analysis.Rmd</code></a></small>
      <div class="d-none name"><code>Vignette_VisiumHD_Mouse_Kidney_analysis.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This vignette demonstrates a full analysis workflow using SpotGLM for
cell type-specific niche differential expression analysis on a Visium HD
data set. We will also demonstrate how <a href="https://kaishumason.github.io/SPARROW/" class="external-link">SPARROW</a> can be used
for power-preserving data selection, to speed up the analysis.</p>
<p>We will use this <a href="https://www.10xgenomics.com/datasets/visium-hd-cytassist-gene-expression-libraries-of-mouse-kidney" class="external-link">publicly
available Visium HD mouse kidney dataset</a>, which can be downloaded
form 10X Genomics.</p>
<p>All runtimes reported are on a Windows 11 (64-bit) machine with 64GB
ram and Intel(R) Core(TM) Ultra 7 165U 1700 Mhz processor with 12 cores.
The total runtime for this script is &lt;30 minutes. The parts of the
script where runtime is not reported should finish instantaneously.</p>
</div>
<div class="section level2">
<h2 id="setup-and-preparing-data-matrices">Setup and preparing data matrices<a class="anchor" aria-label="anchor" href="#setup-and-preparing-data-matrices"></a>
</h2>
<p>First, load some required packages.</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kaishumason/spotGLM">spotGLM</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span> <span class="co"># for plots</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span> <span class="co"># for plots</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggrepel.slowkow.com/" class="external-link">ggrepel</a></span><span class="op">)</span> <span class="co"># for labeling points in qqplot</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span> <span class="co"># for fast manipulation of gene tables</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span></code></pre>
<p>SpotGLM requires three input matrices”</p>
<ul>
<li>
<strong>Spot-by-gene counts matrix</strong>: This is the raw count
matrix directly output by by 10X</li>
<li>
<strong>Spot-by-cell type compositions matrix</strong>: This is the
spot deconvolution result, giving the contribution of each cell type to
each spot. We will use <a href="https://www.nature.com/articles/s41587-021-00830-w" class="external-link">RCTD</a> as an
example, but you can use any other method. Deconvolution can take <em>a
few hours</em>, so we provide this matrix, and you can skip this step
for now.<br>
</li>
<li>
<strong>Spot-by-niche covariate matrix</strong>: This contains the
spatial niche covariates for the right-hand-side of the spotGLM
regression. For this analysis, we will use kidney anatomical regions
defined via clustering of histological image features extracted from the
same-slide H&amp;E image. This feature extraction and clustering takes
<em>a few hours with GPU</em>, so we provide this matrix for you to use.
You can define spatial niches/domains/neighborhoods using other methods
as well.</li>
</ul>
<p>If you would like to try these pre-processing steps yourself, you can
use the script we provide below. Otherwise, skip directly to the next
section.</p>
<details><summary><strong>Pre-processing script for preparing the data matrices.</strong>
</summary><p>We load some libraries used specifically by RCTD.</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/apache/arrow/" class="external-link">arrow</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">spacexr</span><span class="op">)</span> </span></code></pre>
<p>We will also load Seurat, which is used to read in the Visium files
and extract the counts and spatial coordinates.</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span></code></pre>
<p>As described in our paper, we bin the Visium HD data into 8 micron
squares. This alleviates the extreme sparsity in the data. You can
download the binned data from <strong>here</strong>, as well as the cell
type reference file for deconvolution using RCTD. Simply download the
zip file, extract all files into a folder named
“mouse_kidney_data_input”, and put this foler into the root directory
where you are doing the analysis.</p>
<pre><code><span><span class="va">localdir</span> <span class="op">&lt;-</span> <span class="st">"mouse_kidney_data_input/data/binned_outputs/square_008um"</span></span>
<span><span class="va">output_dir</span> <span class="op">&lt;-</span> <span class="st">"mouse_kidney_data_input/preprocessed_data/"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">output_dir</span>, showWarnings <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre>
<p>Read in the data, and run RCTD. This will take a while.</p>
<pre><code><span><span class="va">reference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span> <span class="st">"mouse_kidney_data_input/RCTD_reference_data/Mouse_kidney_RCTD_reference.rds"</span><span class="op">)</span></span>
<span><span class="va">counts</span><span class="op">&lt;-</span><span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X_h5.html" class="external-link">Read10X_h5</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">localdir</span>, <span class="st">"filtered_feature_bc_matrix.h5"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">coords</span><span class="op">&lt;-</span><span class="fu"><a href="https://arrow.apache.org/docs/r/reference/read_parquet.html" class="external-link">read_parquet</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">localdir</span>, <span class="st">"spatial/tissue_positions.parquet"</span><span class="op">)</span>, as_data_frame <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">coords</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">coords</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">coords</span><span class="op">)</span><span class="op">&lt;-</span><span class="va">coords</span><span class="op">$</span><span class="va">barcode</span></span>
<span><span class="va">coords</span> <span class="op">&lt;-</span> <span class="va">coords</span><span class="op">[</span><span class="va">coords</span><span class="op">$</span><span class="va">barcode</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">coords</span><span class="op">&lt;-</span><span class="va">coords</span><span class="op">[</span>,<span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span>
<span><span class="va">nUMI</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">puck</span> <span class="op">&lt;-</span> <span class="fu">SpatialRNA</span><span class="op">(</span><span class="va">coords</span>, <span class="va">counts</span>, <span class="va">nUMI</span><span class="op">)</span></span>
<span><span class="va">barcodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">puck</span><span class="op">@</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">myRCTD</span> <span class="op">&lt;-</span> <span class="fu">create.RCTD</span><span class="op">(</span><span class="va">puck</span>, <span class="va">reference</span>, max_cores <span class="op">=</span> <span class="fl">12</span><span class="op">)</span></span>
<span><span class="va">myRCTD</span> <span class="op">&lt;-</span> <span class="fu">run.RCTD</span><span class="op">(</span><span class="va">myRCTD</span>, doublet_mode <span class="op">=</span> <span class="st">'doublet'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">myRCTD</span>,file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">output_dir</span>,<span class="st">"Deconvolution_HD.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>Now, take the RCTD output, and make the spot composition matrix for
input into spotGLM:</p>
<pre><code><span><span class="va">cell_types</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">reference</span><span class="op">@</span><span class="va">cell_types</span><span class="op">)</span></span>
<span><span class="va">cell_types</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/droplevels.html" class="external-link">droplevels</a></span><span class="op">(</span><span class="va">cell_types</span><span class="op">)</span></span>
<span><span class="va">types</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cell_types</span><span class="op">)</span></span>
<span><span class="va">coords</span> <span class="op">&lt;-</span>  <span class="va">myRCTD</span><span class="op">@</span><span class="va">spatialRNA</span><span class="op">@</span><span class="va">coords</span></span>
<span><span class="va">deconv_est</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">coords</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">types</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">deconv_est</span><span class="op">)</span> <span class="op">=</span> <span class="va">types</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">deconv_est</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">myRCTD</span><span class="op">@</span><span class="va">spatialRNA</span><span class="op">@</span><span class="va">coords</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">myRCTD</span><span class="op">@</span><span class="va">results</span><span class="op">$</span><span class="va">results_df</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">fills</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">myRCTD</span><span class="op">@</span><span class="va">results</span><span class="op">$</span><span class="va">results_df</span><span class="op">[</span><span class="va">j</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"first_type"</span>, <span class="st">"second_type"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, <span class="va">types</span><span class="op">)</span></span>
<span>  <span class="va">deconv_est</span><span class="op">[</span><span class="va">j</span>,<span class="va">fills</span><span class="op">]</span> <span class="op">=</span> <span class="va">myRCTD</span><span class="op">@</span><span class="va">results</span><span class="op">$</span><span class="va">weights_doublet</span><span class="op">[</span><span class="va">j</span>,<span class="op">]</span></span>
<span><span class="op">}</span></span>
<span><span class="va">deconv_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">deconv_est</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">deconv_df</span> , <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">output_dir</span>, <span class="st">"deconv_matrix.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>Write a matrix containing the spatial coordinates:</p>
<pre><code><span><span class="va">coords_all</span><span class="op">&lt;-</span><span class="fu"><a href="https://arrow.apache.org/docs/r/reference/read_parquet.html" class="external-link">read_parquet</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">localdir</span>, <span class="st">"spatial/tissue_positions.parquet"</span><span class="op">)</span>,as_data_frame <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">coords_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">coords_all</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">coords_all</span><span class="op">)</span><span class="op">&lt;-</span><span class="va">coords_all</span><span class="op">$</span><span class="va">barcode</span></span>
<span><span class="va">coord_df</span> <span class="op">&lt;-</span> <span class="va">coords_all</span><span class="op">[</span><span class="va">coords_all</span><span class="op">$</span><span class="va">barcode</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">deconv_df</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">coord_df</span> <span class="op">&lt;-</span> <span class="va">coord_df</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">deconv_df</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">coord_df</span> <span class="op">&lt;-</span> <span class="va">coord_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">barcode</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">coord_df</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">coord_df</span> , <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">output_dir</span>, <span class="st">"coords.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>Write a matrix containing the counts:</p>
<pre><code><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Load10X_Spatial.html" class="external-link">Load10X_Spatial</a></span><span class="op">(</span>data.dir <span class="op">=</span> <span class="va">localdir</span><span class="op">)</span></span>
<span><span class="va">count</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">Spatial</span><span class="op">$</span><span class="va">counts</span></span>
<span><span class="va">count</span> <span class="op">&lt;-</span> <span class="va">count</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">deconv_df</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">count</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">output_dir</span>, <span class="st">"count_matrix.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</details>
</div>
<div class="section level2">
<h2 id="load-data-matrices-and-initial-data-clean">Load data matrices and initial data clean<a class="anchor" aria-label="anchor" href="#load-data-matrices-and-initial-data-clean"></a>
</h2>
<p>We first load the data and prepare for spotGLM, defining the
following variables:</p>
<ul>
<li>
<code>deconv</code>: The deconvolution of each spot, of dimension
number of spots by cell types</li>
<li>
<code>EN</code>: The spatial niche vector (in this case, mouse
kidney regions) for each spot, of dimension number of spots by kidney
regions</li>
<li>
<code>coords</code>: Coordinate matrix, of dimension number of spots
by 2</li>
<li>
<code>counts</code>: The counts matrix for spot-by-gene matrix, of
dimension number of spots by number of genes</li>
</ul>
<pre><code><span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="st">"mouse_kidney_data_input/preprocessed_data/"</span></span>
<span><span class="va">deconv</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>,<span class="st">"deconv_matrix.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">EN</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>,<span class="st">"EN_covariate_matrix.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">coords</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"coords.rds"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>,<span class="fl">2</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="va">counts</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>,<span class="st">"count_matrix.rds"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>For this data, some of the spots didn’t have anatomical annotation.
Thus, we keep only those spots that are in the <code>EN</code>
environment covariate matrix.</p>
<pre><code><span><span class="va">coords</span> <span class="op">=</span> <span class="va">coords</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">EN</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">counts</span> <span class="op">=</span> <span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">EN</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">deconv</span> <span class="op">=</span> <span class="va">deconv</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">EN</span><span class="op">)</span>, <span class="op">]</span></span></code></pre>
<p>To get a sense of what you are working with, you can now visualize
this data. First, let’s plot the spatial niches (the <code>EN</code>
spatial covariate matrix).</p>
<pre><code><span><span class="co"># Convert one-hot matrix to factor of region names</span></span>
<span><span class="va">region_vector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">EN</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/maxCol.html" class="external-link">max.col</a></span><span class="op">(</span><span class="va">EN</span>, ties.method <span class="op">=</span> <span class="st">"first"</span><span class="op">)</span><span class="op">]</span>,levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">EN</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create plotting data frame</span></span>
<span><span class="va">plot_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">coords</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">coords</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  region <span class="op">=</span> <span class="va">region_vector</span>,</span>
<span>  Endothelial <span class="op">=</span> <span class="va">deconv</span><span class="op">[</span>,<span class="st">"Endothelial cells"</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span><span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,y<span class="op">=</span><span class="va">y</span>,fill<span class="op">=</span><span class="va">region</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_raster</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Set1"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"kidney_niches.png"</span>, plot<span class="op">=</span><span class="va">p1</span>, width<span class="op">=</span><span class="fl">4</span>, height<span class="op">=</span><span class="fl">4</span>, dpi<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span></code></pre>
<p><img src="../reference/figures/kidney_niches.png" alt="Kidney spatial regions" style="width:80%;"></p>
<p>Next, let’s plot the distribution of endothelial cells (the
Endothelial cell column of the <code>deconv</code> matrix). Later, we
will be finding endothelial cell-specific niche markers.</p>
<pre><code><span><span class="va">p1</span><span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_df</span><span class="op">[</span>,<span class="op">]</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, fill <span class="op">=</span> <span class="va">Endothelial</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_raster</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"kidney_endothelial.png"</span>, plot<span class="op">=</span><span class="va">p1</span>, width<span class="op">=</span><span class="fl">4</span>, height<span class="op">=</span><span class="fl">4</span>, dpi<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span></code></pre>
<p><img src="../reference/figures/kidney_endothelial.png" alt="Kidney spatial regions" style="width:80%;"></p>
<p>Next, let’s try plotting a gene, Slc5a12:</p>
<pre><code><span><span class="va">plot_df</span><span class="op">$</span><span class="va">Slc5a12</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">1</span><span class="op">+</span><span class="va">counts</span><span class="op">[</span>,<span class="st">"Slc5a12"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">p1</span><span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_df</span><span class="op">[</span>,<span class="op">]</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, fill <span class="op">=</span> <span class="va">Slc5a12</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_raster</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"kidney_Slc5a12.png"</span>, plot<span class="op">=</span><span class="va">p1</span>, width<span class="op">=</span><span class="fl">4</span>, height<span class="op">=</span><span class="fl">4</span>, dpi<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span></code></pre>
<p><img src="../reference/figures/kidney_Slc5a12.png" alt="Kidney spatial regions" style="width:80%;"></p>
<p>The deconvolved cell type proportions sometimes have cell types
contributing very small amounts to a spot We denoise by setting the
proportions of these to 0.</p>
<pre><code><span><span class="va">deconv</span><span class="op">[</span><span class="va">deconv</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">]</span> <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="va">deconv</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">deconv</span>,<span class="fl">1</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="va">x</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>SpotGLM is a gene-specific analysis and can be parallelized across
genes. For this tutorial, we show spotGLM on a small list of genes, so
that the analysis can be completed in a few minutes. We also provide the
parallelization script so that you can run it on the full
transcriptome.</p>
<p>We always recommend filtering out lowly expressed genes, as such
genes increase the number of tests you will need to adjust for, and
reduces the power for other genes. Filtering them out also avoids
unnecessary computation.</p>
<pre><code><span><span class="va">genelist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>,<span class="st">"genelist.rds"</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">counts</span> <span class="op">=</span> <span class="va">counts</span><span class="op">[</span>, <span class="va">genelist</span><span class="op">]</span></span></code></pre>
<p>Finally, put the gene counts, cell type proportions, and niche
environment all into one <code>list</code> object for passing to our
functions.</p>
<pre><code><span><span class="va">data</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>niche<span class="op">=</span><span class="va">EN</span>, deconv<span class="op">=</span><span class="va">deconv</span>, counts<span class="op">=</span><span class="va">counts</span>, coords<span class="op">=</span><span class="va">coords</span><span class="op">)</span></span></code></pre>
<p>Next, we use SPARROW to reduce the data to a smaller size while
preserving power for this analysis. This is optional, and you can skip
to the next section if you want to simply run on the full data set.</p>
</div>
<div class="section level2">
<h2 id="use-sparrow-for-power-preserving-data-reduction">Use SPARROW for power-preserving data reduction<a class="anchor" aria-label="anchor" href="#use-sparrow-for-power-preserving-data-reduction"></a>
</h2>
<p><a href="https://kaishumason.github.io/SPARROW/" class="external-link">SPARROW</a> is a
method that selects a small representative sub-sample from a large
dataset to reduce computation time. The sub-sample is chosen to maximize
the power of a downstream analysis, which, in this case, is SpotGLM with
the given spatial covariates and cell type decomposition. SPARROW
computes the number of spots needed for high power, and chooses the
spots to get balanced representation across cell type-niche
configurations.</p>
<p>Here, as demonstration, we will use sparrow to select 10,000 spots
with 18 cell types and 7 spatial niches. For a full data analysis, you
should let SPARROW automatically select the number of spots based on
power calculations. The compute time of this part of the script takes
about 2 minutes, but if you were to remove the 10,000 spot restriction,
it will take about 20 minutes. However, this gene selection step can cut
the run-time of the spotGLM analysis from 1 week to overnight, with
minimal loss of power.</p>
<p>First, create the expanded covariate matrix, as described in our
paper.</p>
<pre><code><span><span class="va">cutoff</span> <span class="op">=</span> <span class="op">-</span><span class="fl">6.5</span></span>
<span><span class="va">expanded_X</span> <span class="op">=</span> <span class="fu">sparrow</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/sparrow/man/expand_covariate_matrix.html" class="external-link">expand_covariate_matrix</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">niche</span>, lambda <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">deconv</span>,</span>
<span>                                              family <span class="op">=</span> <span class="st">"negative binomial"</span>,lib_size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>,min_reads_per_1000 <span class="op">=</span> <span class="fl">1000</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">cutoff</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>You can take a look at the columns of expanded_X, to get a sense of
what this matrix holds. It gives the weight of each spot in every
possible (cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>×</mi><annotation encoding="application/x-tex">\times</annotation></semantics></math>
spatial niche) combination.</p>
<p>Next, remove the covariates with too many zero entries, and compute
the target standard error. The target standard error is the standard
error needed to achieve a power of <code>target_power_approx</code>
(here set to <code>99\%</code>) on effects of size
<code>min_effect</code> (here set to 0.05). These target standard errors
will be used in the next step to determine how much data reduction we
can tolerate, and which spots we should select, to guarantee the desired
power.</p>
<pre><code><span><span class="va">freq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">expanded_X</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="va">bad_cov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">freq</span> <span class="op">&lt;</span> <span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">expanded_X</span> <span class="op">&lt;-</span> <span class="va">expanded_X</span><span class="op">[</span>, <span class="op">-</span><span class="va">bad_cov</span><span class="op">]</span></span>
<span><span class="va">target_standard_errors</span> <span class="op">=</span> <span class="fu">sparrow</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/sparrow/man/compute_target_standard_error.html" class="external-link">compute_target_standard_error</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">expanded_X</span>,min_effect <span class="op">=</span> <span class="fl">0.05</span>,target_power_approx <span class="op">=</span> <span class="fl">0.99</span><span class="op">)</span></span></code></pre>
<p>The <code>data_selection</code> function in Sparrow selects the spots
based on the expanded covariate matrix and the target standard errors.
Here, we set a maximum of 10,000 spots. In a full data analysis, we
recommend increasing this maximum to 100,000 spots.</p>
<pre><code><span><span class="va">selected_indices</span> <span class="op">=</span> <span class="fu">sparrow</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/sparrow/man/data_selection.html" class="external-link">data_selection</a></span><span class="op">(</span>X <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">expanded_X</span><span class="op">)</span>,</span>
<span>                                           max_data_size <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>                                           min_standard_error <span class="op">=</span> <span class="va">target_standard_errors</span>,</span>
<span>                                           log <span class="op">=</span> <span class="cn">TRUE</span>,period <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span>
<span><span class="va">selected_indices</span> <span class="op">=</span> <span class="va">selected_indices</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">selected_indices</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"#Cells Subsampled: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">selected_indices</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>You can visualize which indices were selected, overlayed on the
spatial map of the anatomical niches:</p>
<pre><code><span><span class="va">p1</span><span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,y<span class="op">=</span><span class="va">y</span>,fill<span class="op">=</span><span class="va">region</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_raster</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Set1"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># overlay with selected indices</span></span>
<span><span class="va">overlay_df</span> <span class="op">&lt;-</span> <span class="va">plot_df</span><span class="op">[</span><span class="va">selected_indices</span>, <span class="op">]</span></span>
<span><span class="va">p1</span><span class="op">=</span><span class="va">p1</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">overlay_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span><span class="op">)</span>,</span>
<span>             color <span class="op">=</span> <span class="st">"black"</span>, shape <span class="op">=</span> <span class="fl">21</span>, size <span class="op">=</span> <span class="fl">1.2</span>, stroke <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"kidney_selected_indices.png"</span>, plot<span class="op">=</span><span class="va">p1</span>, width<span class="op">=</span><span class="fl">8</span>, height<span class="op">=</span><span class="fl">8</span>, dpi<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span></code></pre>
<p><img src="../reference/figures/kidney_selected_indices.png" alt="Kidney spatial regions" style="width:80%;"></p>
<p>We see that the selected spots are distributed across the anatomic
regions, and all regions are represented. However, the spots are not
uniformly distributed but are more dense in some regions; that is
because those regions have higher complexity in cell types or cell
type-niche combinations. If we were to explore the deconvolved cell
types proportions within each niche, we would find that every cell
type-niche combination is also evenly represented. This allows us to
retain the desired power in differential tests involving all of these
cell types and spatial niches.</p>
<p>Now, we can subset the data to the selected indices for downstream
analysis</p>
<pre><code><span><span class="va">data</span><span class="op">$</span><span class="va">deconv</span> <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">[</span><span class="va">selected_indices</span>,<span class="op">]</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">counts</span> <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="va">selected_indices</span>,<span class="op">]</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">coords</span> <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span><span class="va">selected_indices</span>,<span class="op">]</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">niche</span> <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">niche</span><span class="op">[</span><span class="va">selected_indices</span>,<span class="op">]</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">library_size</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="fit-spotglm-model-and-identify-cell-type-specific-niche-markers">Fit SpotGLM model and identify cell type-specific niche markers<a class="anchor" aria-label="anchor" href="#fit-spotglm-model-and-identify-cell-type-specific-niche-markers"></a>
</h2>
<p>We are now ready to run spotGLM on the selected data. This proceeds
in two steps. First, we fit a spotglm model (using
<code><a href="../reference/run_model.html">spotGLM::run_model</a></code>) on each gene to estimate its cell
type-specific coefficients for each spatial covariate. This is the most
time-consuming step of the analysis, which <em>should be parallelized
across genes</em>. After fitting the model, you can then test all sorts
of hypotheses with it, such as contrasting two spatial niches to
identify cell type-specific niche markers.</p>
<details><summary>
Expand to see the model that we are fitting.
</summary><p>Let
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
be the the covariate matrix where the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>-th
row is the vector describing the spatial domain of spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>.
In this example, the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>-th
row has dimension 7, for 7 anatomic domains, and is the one-hot encoding
of the spot’s anatomic domain. We model the expression of a gene in spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>
as Poisson, with mean that is dependent on the spatial covariates:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Y</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi></mrow></msub><mo>∼</mo><mi>P</mi><mi>o</mi><mi>i</mi><mi>s</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝔼</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>Y</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi></mrow></msub><mo stretchy="false" form="prefix">|</mo><msub><mi>X</mi><mi>s</mi></msub><mo stretchy="true" form="postfix">]</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">Y_{s,g} \sim Pois(\mathbb{E}[Y_{s,g}|X_{s}])</annotation></semantics></math>
The mean expression is equal to the sum of contributions across cell
types that belong to the spot,</p>
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>𝔼</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>Y</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi></mrow></msub><mo stretchy="false" form="prefix">|</mo><msub><mi>X</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">]</mo></mrow><mo>=</mo><mo>log</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>L</mi><mi>s</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><munder><mo>∑</mo><mi>t</mi></munder><msub><mi>π</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><mo>exp</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>X</mi><mi>s</mi></msub><msubsup><mi>β</mi><mi>g</mi><mi>t</mi></msubsup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbb{E}[Y_{s,g}|X_{i}] = \log(L_s) + \sum_{t} \pi_{s,t}\exp(X_s\beta^{t}_g) </annotation></semantics></math>
Here
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>L</mi><mi>s</mi></msub><annotation encoding="application/x-tex">L_s</annotation></semantics></math>
is the library size of spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>π</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">\pi_{s,t}</annotation></semantics></math>
is the deconvolution estimate for cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
in spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>.
The contribution of each cell type depends on the spatial covariates
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>X</mi><mi>s</mi></msub><annotation encoding="application/x-tex">X_s</annotation></semantics></math>
through the parameter vector
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>β</mi><mi>g</mi><mi>t</mi></msubsup><annotation encoding="application/x-tex">\beta_g^t</annotation></semantics></math>.
This vector has dimension equal to the number of columns in
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>.
During model fitting, we are estimating
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>β</mi><mi>g</mi><mi>t</mi></msubsup><annotation encoding="application/x-tex">\beta_g^t</annotation></semantics></math>
for each gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
and each cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>.
Downstream, the estimates and standard errors of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>β</mi><mi>g</mi><mi>t</mi></msubsup><annotation encoding="application/x-tex">\beta_g^t</annotation></semantics></math>
will be used for cell type-specific niche-differential expression
analysis.
</details><div class="section level3">
<h3 id="testing-on-a-reduced-set-of-genes">Testing on a reduced set of genes<a class="anchor" aria-label="anchor" href="#testing-on-a-reduced-set-of-genes"></a>
</h3>
<p>In this tutorial, we first test the model for ~230 genes. This
process takes ~15 minutes.</p>
<pre><code><span><span class="va">t1</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">num_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">spotglm_fit</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="st">"list"</span>,<span class="va">num_genes</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">num_genes</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="va">j</span><span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span><span class="fl">5</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"fitting model for gene"</span>, <span class="va">j</span>, <span class="st">"out of"</span>,<span class="va">num_genes</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">t1</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">spotglm_fit</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu">spotGLM</span><span class="fu">::</span><span class="fu"><a href="../reference/run_model.html">run_model</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span>,<span class="va">j</span><span class="op">]</span>, X <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">niche</span>, lambda <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">deconv</span>,</span>
<span>                             family <span class="op">=</span> <span class="st">"spot negative binomial"</span>, offset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">library_size</span><span class="op">)</span>,</span>
<span>                             batch_size <span class="op">=</span> <span class="fl">250</span>, learning_rate <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"done!"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">spotglm_fit</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
<p>To run spotGLM on all genes, we recommend using parallel computing
library <code>parallel</code>. Please expand to get the script for doing
this.</p>
<details><summary>
spotGLM with parallelization across all genes
</summary><pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">parallel</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RevolutionAnalytics/foreach" class="external-link">foreach</a></span><span class="op">)</span></span>
<span><span class="va">family</span> <span class="op">&lt;-</span> <span class="st">"spot negative binomial"</span></span>
<span><span class="va">G</span> <span class="op">=</span> <span class="fl">0.1</span></span>
<span><span class="va">data_size</span> <span class="op">=</span> <span class="fl">8</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/prod.html" class="external-link">prod</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fl">1e+09</span></span>
<span><span class="co">#number of chunks needed</span></span>
<span><span class="va">nchunks</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="va">data_size</span><span class="op">/</span><span class="va">G</span><span class="op">)</span></span>
<span><span class="va">chunk_size</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">/</span><span class="va">nchunks</span><span class="op">)</span></span>
<span><span class="va">grouping</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">nchunks</span>, each <span class="op">=</span> <span class="va">chunk_size</span>, length.out <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">index_chunks</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>, <span class="va">grouping</span><span class="op">)</span></span>
<span><span class="va">chunk_counter</span> <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="va">T_1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">I</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">index_chunks</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">t1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Initializing cluster"</span><span class="op">)</span></span>
<span>  <span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></span><span class="op">(</span><span class="fl">5</span>, outfile <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span>  <span class="fu">doParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/doParallel/man/registerDoParallel.html" class="external-link">registerDoParallel</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/parallel/clusterApply.html" class="external-link">clusterExport</a></span><span class="op">(</span><span class="va">cluster</span>, varlist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"spot_negative_binomial"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Evaluating chunk "</span>, <span class="va">I</span>, <span class="st">" out of "</span>,</span>
<span>               <span class="va">nchunks</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">counts_chunk</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">counts</span><span class="op">[</span>, <span class="va">index_chunks</span><span class="op">[[</span><span class="va">I</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">NC</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">counts_chunk</span><span class="op">)</span></span>
<span>  <span class="va">results_chunk</span> <span class="op">=</span> <span class="fu">foreach</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/foreach/man/foreach.html" class="external-link">foreach</a></span><span class="op">(</span>i <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">NC</span>, .export <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X"</span>, <span class="st">"lambda"</span>, <span class="st">"family"</span>,<span class="st">"offset"</span>,</span>
<span>                                                         <span class="st">"counts_chunk"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/foreach/man/foreach.html" class="external-link">%dopar%</a></span> <span class="op">{</span></span>
<span>                                                           <span class="kw">if</span><span class="op">(</span><span class="va">i</span><span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span><span class="fl">1</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span><span class="op">{</span></span>
<span>                                                             <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span></span>
<span>                                                             <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span>                                                           <span class="op">}</span></span>
<span>                                                           </span>
<span>                                                           <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span>                                                           <span class="va">Y</span> <span class="op">=</span> <span class="va">counts_chunk</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span></span>
<span>                                                           <span class="fu">spotglm</span><span class="fu">::</span><span class="fu">run_model</span><span class="op">(</span><span class="va">Y</span>, X <span class="op">=</span> <span class="va">X</span>, lambda <span class="op">=</span> <span class="va">lambda</span>,family <span class="op">=</span> <span class="va">family</span>, offset <span class="op">=</span> <span class="va">offset</span>,learning_rate <span class="op">=</span> <span class="fl">0.1</span>, batch_size <span class="op">=</span> <span class="fl">128</span>,n_epoch <span class="op">=</span> <span class="fl">250</span><span class="op">)</span></span>
<span>                                                         <span class="op">}</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">chunk_counter</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">spotglm_fit</span> <span class="op">=</span> <span class="va">results_chunk</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">spotglm_fit</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">spotglm_fit</span>, <span class="va">results_chunk</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">counts_chunk</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/gc.html" class="external-link">gc</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="va">chunk_counter</span> <span class="op">=</span> <span class="va">chunk_counter</span> <span class="op">+</span> <span class="fl">1</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Closing cluster"</span><span class="op">)</span></span>
<span>  <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">stopCluster</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Chunk took "</span>,<span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">t1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">spotglm_fit</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">T_2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">T_1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">T_2</span><span class="op">)</span></span></code></pre>
</details><p>Now that you have fit the spotglm model on each gene, you can
identify cell type-specific gene markers for spatial niches.</p>
</div>
</div>
<div class="section level2">
<h2 id="identifying-cell-type-specific-niche-marker-genes">Identifying cell type-specific niche marker genes<a class="anchor" aria-label="anchor" href="#identifying-cell-type-specific-niche-marker-genes"></a>
</h2>
<p>To identify niche marker genes, we perform a <em>contrast test</em>,
contrasting the cell type-specific niche coefficients for each gene,
between spatial niches of interest. We can do this using the
<code><a href="../reference/compute_contrast_significance.html">compute_contrast_significance()</a></code> function. We need to tell
this function which spatial covariates (i.e. niches) we are
contrasting.</p>
<details>
A niche marker gene for a cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
is one in which
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>β</mi><mrow><msub><mi>n</mi><mn>1</mn></msub><mo>,</mo><mi>g</mi></mrow><mi>t</mi></msubsup><mo>≠</mo><msubsup><mi>β</mi><mrow><msub><mi>n</mi><mn>2</mn></msub><mo>,</mo><mi>g</mi></mrow><mi>t</mi></msubsup></mrow><annotation encoding="application/x-tex">\beta^{t}_{n_1,g} \neq \beta^{t}_{n_2,g}</annotation></semantics></math>
that is that we can differentiate whether cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
is near a niche enriched by niche1
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>n</mi><mn>1</mn></msub><annotation encoding="application/x-tex">n_1</annotation></semantics></math>
or niche
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>n</mi><mn>2</mn></msub><annotation encoding="application/x-tex">n_2</annotation></semantics></math>
by looking at its gene expression for gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>.
</details><p>For example, let’s find the niche marker genes for differentiating
Endothelial cells in the Papilla region vs those in the Cortex region.
For papilla region markers, we want genes that are upregulated in the
papilla region. Thus, we specify that we want a one-sided test in the
positive direction (side = 1, direction = “pos”).</p>
<pre><code><span><span class="va">cell_type</span> <span class="op">=</span> <span class="st">"Endothelial cells"</span></span>
<span><span class="va">niche1</span> <span class="op">=</span> <span class="st">".data_Papilla"</span> </span>
<span><span class="va">niche2</span> <span class="op">=</span> <span class="st">".data_Cortex"</span></span>
<span><span class="va">sig_genes</span> <span class="op">=</span> <span class="fu"><a href="../reference/compute_contrast_significance.html">compute_contrast_significance</a></span><span class="op">(</span>input_list <span class="op">=</span> <span class="va">spotglm_fit</span>,</span>
<span>                                          cell_type <span class="op">=</span> <span class="va">cell_type</span>,</span>
<span>                                          effect_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">niche1</span>, <span class="va">niche2</span><span class="op">)</span>,</span>
<span>                                          beta_name <span class="op">=</span> <span class="st">"beta_estimate"</span>,covariance_name <span class="op">=</span> <span class="st">"vcov"</span>,</span>
<span>                                          sided <span class="op">=</span> <span class="fl">1</span>,direction <span class="op">=</span> <span class="st">"pos"</span><span class="op">)</span></span>
<span><span class="va">sig_genes</span> <span class="op">=</span> <span class="va">sig_genes</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">pval</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="data-visualization">Data Visualization<a class="anchor" aria-label="anchor" href="#data-visualization"></a>
</h2>
<div class="section level3">
<h3 id="qqplot">QQplot<a class="anchor" aria-label="anchor" href="#qqplot"></a>
</h3>
<p>First, we use a QQ plot to get a sense of what the p-value
distribution look like, and highlight the significant genes.</p>
<pre><code><span><span class="va">sig_genes_df</span> <span class="op">&lt;-</span> <span class="va">sig_genes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">sig_genes</span><span class="op">$</span><span class="va">test_statistic</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">sig_genes_df</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sig_genes_df</span><span class="op">)</span></span>
<span></span>
<span><span class="va">qq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/qqnorm.html" class="external-link">qqnorm</a></span><span class="op">(</span><span class="va">sig_genes_df</span><span class="op">$</span><span class="va">test_statistic</span>, plot.it <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">sig_genes_df</span><span class="op">$</span><span class="va">theoretical</span> <span class="op">&lt;-</span> <span class="va">qq</span><span class="op">$</span><span class="va">x</span>  <span class="co"># theoretical quantiles</span></span>
<span><span class="va">sig_genes_df</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="va">qq</span><span class="op">$</span><span class="va">y</span>       <span class="co"># sample quantiles</span></span>
<span><span class="co"># Compute the slope and intercept for the QQ line</span></span>
<span><span class="va">q1_sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">sig_genes_df</span><span class="op">$</span><span class="va">test_statistic</span>, <span class="fl">0.25</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">q3_sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">sig_genes_df</span><span class="op">$</span><span class="va">test_statistic</span>, <span class="fl">0.75</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">q1_theor</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">qnorm</a></span><span class="op">(</span><span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="va">q3_theor</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">qnorm</a></span><span class="op">(</span><span class="fl">0.75</span><span class="op">)</span></span>
<span><span class="va">slope</span>     <span class="op">&lt;-</span> <span class="op">(</span><span class="va">q3_sample</span> <span class="op">-</span> <span class="va">q1_sample</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">q3_theor</span> <span class="op">-</span> <span class="va">q1_theor</span><span class="op">)</span></span>
<span><span class="va">intercept</span> <span class="op">&lt;-</span> <span class="va">q1_sample</span> <span class="op">-</span> <span class="va">slope</span> <span class="op">*</span> <span class="va">q1_theor</span></span>
<span></span>
<span><span class="co"># Select the top genes by highest t_test_stat</span></span>
<span><span class="va">top_n</span> <span class="op">&lt;-</span> <span class="fl">35</span></span>
<span><span class="va">top_df</span> <span class="op">&lt;-</span> <span class="va">sig_genes_df</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">sig_genes_df</span><span class="op">$</span><span class="va">test_statistic</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">top_n</span><span class="op">]</span>, <span class="op">]</span></span>
<span><span class="co"># Create the QQ plot using ggplot2 with ggrepel for labeled boxes</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">sig_genes_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">theoretical</span>, y <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>intercept <span class="op">=</span> <span class="va">intercept</span>, slope <span class="op">=</span> <span class="va">slope</span>, color <span class="op">=</span> <span class="st">"red"</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggrepel.slowkow.com/reference/geom_text_repel.html" class="external-link">geom_label_repel</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">top_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">gene</span><span class="op">)</span>,</span>
<span>                   box.padding <span class="op">=</span> <span class="fl">0.35</span>, point.padding <span class="op">=</span> <span class="fl">0.5</span>, <span class="co"># size = 10, </span></span>
<span>                   segment.color <span class="op">=</span> <span class="st">"grey50"</span>, max.overlaps <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Theoretical Quantiles"</span>, </span>
<span>       y <span class="op">=</span> <span class="st">"Sample Quantiles"</span>, </span>
<span>       title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">cell_type</span>, <span class="st">": "</span>, <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace</a></span><span class="op">(</span><span class="va">niche1</span>, <span class="st">".data_"</span>, <span class="st">""</span><span class="op">)</span>, <span class="st">" vs."</span>, <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace</a></span><span class="op">(</span><span class="va">niche2</span>, <span class="st">".data_"</span>, <span class="st">""</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span>, hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"kidney_qqplot.png"</span>, plot<span class="op">=</span><span class="va">p</span>, width<span class="op">=</span><span class="fl">8</span>, height<span class="op">=</span><span class="fl">8</span>, dpi<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span></code></pre>
<p><img src="../reference/figures/kidney_qqplot.png" alt="Example gene spatial map" style="width:80%;"></p>
</div>
<div class="section level3">
<h3 id="spatial-map-of-example-gene-igfbp3">Spatial map of example gene: Igfbp3<a class="anchor" aria-label="anchor" href="#spatial-map-of-example-gene-igfbp3"></a>
</h3>
<p>We can also visualize the spatial distribution of a specific gene.
First, we normalized the counts: <span class="math display">$$
\hbox{normalized count} = \log\left(1+10000\times
\frac{\hbox{count}}{\hbox{library-size}}\right)$$</span> We didn’t have
to do this until now, because the SPARROW and SpotGLM package internally
accounts for library size in their computations. For plotting, we need
to use the normalized values.</p>
<pre><code><span><span class="va">counts_data</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">counts</span></span>
<span><span class="va">total_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">counts_data</span><span class="op">)</span></span>
<span><span class="va">scale_factor</span> <span class="op">&lt;-</span> <span class="fl">1e4</span></span>
<span><span class="va">scaled_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">/</span> <span class="va">total_counts</span><span class="op">)</span> <span class="op">*</span> <span class="va">scale_factor</span></span>
<span><span class="va">normalized_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">scaled_counts</span><span class="op">)</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">normcounts</span> <span class="op">&lt;-</span> <span class="va">normalized_counts</span></span></code></pre>
<p>Let’s look at a marker that we just identified for distinguishing
between endothelial cells in the Papilla region versus those in the
Cortex region.</p>
<pre><code><span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="st">"Igfbp3"</span></span>
<span></span>
<span><span class="co"># Filter for spots only in the two regions being compared</span></span>
<span><span class="va">region_filter</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">niche</span><span class="op">[</span>, <span class="st">".data_Cortex"</span><span class="op">]</span> <span class="op">==</span> <span class="fl">1</span> <span class="op">|</span> <span class="va">data</span><span class="op">$</span><span class="va">niche</span><span class="op">[</span>, <span class="st">".data_IM"</span><span class="op">]</span> <span class="op">==</span> <span class="fl">1</span></span>
<span><span class="va">gene_expr</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">normcounts</span><span class="op">[</span><span class="va">region_filter</span>, <span class="va">gene</span><span class="op">]</span></span>
<span><span class="va">cellprop_score</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">[</span><span class="va">region_filter</span>, <span class="st">"Endothelial cells"</span> <span class="op">]</span></span>
<span><span class="va">region_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">niche</span><span class="op">[</span><span class="va">region_filter</span>, <span class="st">".data_Cortex"</span><span class="op">]</span> <span class="op">==</span> <span class="fl">1</span>, <span class="st">"Cortex"</span>, <span class="st">"IM"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create plotting data frame</span></span>
<span><span class="va">plot_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data_sparrow</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span><span class="va">region_filter</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">data_sparrow</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span><span class="va">region_filter</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  Gene <span class="op">=</span> <span class="va">gene_expr</span>,</span>
<span>  region <span class="op">=</span> <span class="va">region_label</span>,</span>
<span>  celltype <span class="op">=</span> <span class="va">cellprop_score</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># === Plot 1: Gene expression levels ===</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">Gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span>, alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">gene</span>, <span class="st">" Gene expression value"</span><span class="op">)</span>, color <span class="op">=</span> <span class="va">gene</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradient</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"blue"</span>, high <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># === Plot 2: Region annotation ===</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">region</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span>, alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cortex"</span> <span class="op">=</span> <span class="st">"#1f78b4"</span>, <span class="st">"IM"</span> <span class="op">=</span> <span class="st">"#33a02c"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Region Membership"</span>, color <span class="op">=</span> <span class="st">"Region"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># === Plot 3: Endoehelial cell type proportion ===</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">celltype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span>, alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Deconvolution Proportion"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span>  </span>
<span></span>
<span></span>
<span><span class="va">combined_plot</span> <span class="op">=</span> <span class="va">p1</span><span class="op">+</span><span class="va">p2</span><span class="op">+</span><span class="va">p3</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"example_gene.png"</span>, plot<span class="op">=</span><span class="va">combined_plot</span>, width<span class="op">=</span><span class="fl">15</span>, height<span class="op">=</span><span class="fl">5</span>, dpi<span class="op">=</span><span class="fl">300</span><span class="op">)</span></span></code></pre>
<p><img src="../reference/figures/example_gene.png" alt="Example gene spatial map" style="width:100%;"></p>
<p>We see that while endothelial cells are present in both the cortex
and IM, the endothelial gene <code>Igfbp3</code> has visible expression
enrichment only in the cortex.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kaishu Mason, Yijia Jiang, Nancy R. Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
