<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Visium Analysis • spotGLM</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Visium Analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">spotGLM</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Intro_to_SpotGLM.html">Intro to SpotGLM</a></li>
    <li><a class="dropdown-item" href="../articles/Spatial_ATAC_analysis.html">Spatial ATAC Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Spatial_Long_Read_analysis.html">Spatial Long-Read Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Vignette_VisiumHD_Mouse_Kidney_analysis.html">Visium Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Visium_analysis.html">Visium Analysis</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/kaishumason/spotGLM/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Visium Analysis</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/kaishumason/spotGLM/blob/HEAD/vignettes/Visium_analysis.Rmd"><code>vignettes/Visium_analysis.Rmd</code></a></small>
      <div class="d-none name"><code>Visium_analysis.Rmd</code></div>
    </div>

    
    
<p>We must load the following packages</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">spotglm</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this vignette we will go through a typical analysis using spotGLM
on a Visium dataset. The dataset comes from a liver colorectal cancer
patient. In the dataset, we identify 6 major cell types, hepatocytes,
macrophages/myeloid, fibroblasts/stromal, and two subclones of the tumor
found via <a href="https://www.biorxiv.org/content/10.1101/2022.07.05.498882v2" class="external-link">clonalscope</a>.
For each spot we summarize its neighborhood/niche by computing a
weighted sum of deconvolutions with closer spots having more weight. We
will then fit a spotGLM to identify cell type specific
niche-differential genes.</p>
<div class="section level3">
<h3 id="load-data">Load Data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h3>
<p>We first load the data.</p>
</div>
</div>
<div class="section level2">
<h2 id="reading-in-data">Reading in Data<a class="anchor" aria-label="anchor" href="#reading-in-data"></a>
</h2>
<p>First we read in the merfish data using the function ‘read merfish’.
This functions returns a list with the following</p>
<details><summary>
Output
</summary><ul>
<li>
<code>coords</code>: Coordinate matrix. Of dimension 3826 by 2</li>
<li>
<code>niche</code>: The niche vector for each spot. Of dimension
3826 by 7</li>
<li>
<code>deconv</code>: The deconvolution of each spot. of dimension
3826 by 7</li>
<li>
<code>counts</code>: The counts matrix for 2000 genes. Of dimension
3826 by 2000</li>
<li>
<code>library size</code>: The library size of each spot. A vector
of length 3826</li>
</ul></details><pre><code><span><span class="va">data</span> <span class="op">=</span> <span class="fu">spotglm</span><span class="fu">::</span><span class="fu">read_visium</span><span class="op">(</span><span class="op">)</span></span></code></pre>
<div class="section level3">
<h3 id="example-of-covariate">Example of Covariate<a class="anchor" aria-label="anchor" href="#example-of-covariate"></a>
</h3>
<p>We summarize each spot’s niche into a vector of length
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>T</mi><annotation encoding="application/x-tex">T</annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>T</mi><annotation encoding="application/x-tex">T</annotation></semantics></math>
is the number of cell types in the sample. This covariate is called an
effective niche.Without going into too much detail, the larger the
covariate, the more the corresponding cell type is in the neighborhood
of that spot. For example, we see that the tumor subclone 1 niche
covariate is high in regions where the tumor deconvolution is also
high.</p>
<pre><code><span><span class="co"># Prepare data for plotting</span></span>
<span><span class="va">plot_df_deconv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  subclone_1 <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">[</span>, <span class="st">"subclone_1"</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_df_niche</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  subclone_1 <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">niche</span><span class="op">[</span>, <span class="st">"subclone_1"</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot 1: deconv</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_df_deconv</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">subclone_1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Deconvolution - Subclone 1"</span>, color <span class="op">=</span> <span class="st">"Subclone 1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot 2: niche</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_df_niche</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">subclone_1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Niche Covariate - Subclone 1"</span>, color <span class="op">=</span> <span class="st">"Subclone 1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Display side by side</span></span></code></pre>
<pre><code><span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="perform-spotglm-approx-10-minutes">Perform SpotGLM (Approx 10 minutes)<a class="anchor" aria-label="anchor" href="#perform-spotglm-approx-10-minutes"></a>
</h3>
<p>We now fit our spotGLM model. Let
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>X</mi><annotation encoding="application/x-tex">X</annotation></semantics></math>
be the the covariate matrix where the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>th
row of X,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>X</mi><mi>i</mi></msub><annotation encoding="application/x-tex">X_i</annotation></semantics></math>
corresponds to the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>T</mi><annotation encoding="application/x-tex">T</annotation></semantics></math>
dimensional effective niche vector described above. The single cell glm
model for each cell type is
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Y</mi><mrow><mi>i</mi><mo>,</mo><mi>g</mi></mrow></msub><mo>∼</mo><mi>P</mi><mi>o</mi><mi>i</mi><mi>s</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝔼</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>Y</mi><mrow><mi>i</mi><mo>,</mo><mi>g</mi></mrow></msub><mo stretchy="false" form="prefix">|</mo><msub><mi>X</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">]</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">Y_{i,g} \sim Pois(\mathbb{E}[Y_{i,g}|X_{i}])</annotation></semantics></math><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>log</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝔼</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>Y</mi><mrow><mi>i</mi><mo>,</mo><mi>g</mi></mrow></msub><mo stretchy="false" form="prefix">|</mo><msub><mi>X</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">]</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msub><mi>X</mi><mi>i</mi></msub><msubsup><mi>β</mi><mi>g</mi><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msubsup><mo>+</mo><mo>log</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>L</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\log(\mathbb{E}[Y_{i,g}|X_{i}]) = X_i\beta^{CT(i)}_g + \log(L_i)</annotation></semantics></math>
Here
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>L</mi><mi>i</mi></msub><annotation encoding="application/x-tex">L_i</annotation></semantics></math>
is the library size of cell
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">CT(i)</annotation></semantics></math>
is the cell type assignment of cell
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>β</mi><mi>g</mi><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msubsup><annotation encoding="application/x-tex">\beta^{CT(i)}_g</annotation></semantics></math>
is a vector of dimension
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>T</mi><mo>×</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">T \times 1</annotation></semantics></math>
that acts as a weight vector for cells of type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">CT(i)</annotation></semantics></math>
for gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>.
If an index of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>β</mi><mi>g</mi><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msubsup><annotation encoding="application/x-tex">\beta^{CT(i)}_g</annotation></semantics></math>,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>β</mi><mrow><mi>t</mi><mi>′</mi><mo>,</mo><mi>g</mi></mrow><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msubsup><annotation encoding="application/x-tex">\beta^{CT(i)}_{t',g}</annotation></semantics></math>
is not equal to 0, then the amount of cell
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>t</mi><mi>′</mi></mrow><annotation encoding="application/x-tex">t'</annotation></semantics></math>
in the niche affects the expression of gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
in cells of type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>C</mi><mi>T</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>i</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">CT(i)</annotation></semantics></math>.
This is called a (CT(i),t’) niche differential gene.<br></p>
<p>For a spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>,
its gene expression is modeled as the sum of the cells that make up the
sot and thus can be modeled as</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Y</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi></mrow></msub><mo>∼</mo><mi>P</mi><mi>o</mi><mi>i</mi><mi>s</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>𝔼</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>Y</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi></mrow></msub><mo stretchy="false" form="prefix">|</mo><msub><mi>X</mi><mi>s</mi></msub><mo stretchy="true" form="postfix">]</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">Y_{s,g} \sim Pois(\mathbb{E}[Y_{s,g}|X_{s}])</annotation></semantics></math><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>𝔼</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>Y</mi><mrow><mi>s</mi><mo>,</mo><mi>g</mi></mrow></msub><mo stretchy="false" form="prefix">|</mo><msub><mi>X</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">]</mo></mrow><mo>=</mo><mo>log</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>L</mi><mi>s</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>+</mo><munder><mo>∑</mo><mi>t</mi></munder><msub><mi>π</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><mo>exp</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>X</mi><mi>s</mi></msub><msubsup><mi>β</mi><mi>g</mi><mi>t</mi></msubsup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\mathbb{E}[Y_{s,g}|X_{i}] = \log(L_s) + \sum_{t} \pi_{s,t}\exp(X_s\beta^{t}_g) </annotation></semantics></math>
Here
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>L</mi><mi>s</mi></msub><annotation encoding="application/x-tex">L_s</annotation></semantics></math>
is the library size of spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>π</mi><mrow><mi>s</mi><mo>,</mo><mi>t</mi></mrow></msub><annotation encoding="application/x-tex">\pi_{s,t}</annotation></semantics></math>
is the deconvolution estimate for cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
in spot
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>.
This model can be fit using a spotGLM. to do this we use the “run model”
function. This function takes several arguments, the most important of
which are below This model can be fit using a <code>spotGLM</code>. To
do this, we use the <code><a href="../reference/run_model.html">run_model()</a></code> function.</p>
<pre><code><span><span class="va">num_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="st">"list"</span>,<span class="va">num_genes</span><span class="op">)</span></span>
<span><span class="va">t1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#run spotGLM for each gene</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">num_genes</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="va">j</span><span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span><span class="fl">500</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">j</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">t1</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">res</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu">spotglm</span><span class="fu">::</span><span class="fu">run_model</span><span class="op">(</span>y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span>,<span class="va">j</span><span class="op">]</span>,X <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">niche</span>, lambda <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">deconv</span>,</span>
<span>                             family <span class="op">=</span> <span class="st">"spot poisson"</span>,offset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">library_size</span><span class="op">)</span>,</span>
<span>                             initialization <span class="op">=</span> <span class="cn">T</span>,batch_size <span class="op">=</span> <span class="fl">250</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
<div class="section level4">
<h4 id="downstream-analysis-determining-niche-differential-genes">Downstream Analysis: Determining Niche Differential Genes<a class="anchor" aria-label="anchor" href="#downstream-analysis-determining-niche-differential-genes"></a>
</h4>
<p>After fitting the spotGLM we can evaluate which genes as (t,t’) niche
differential. In this case we get pvalues/qvalues for testing if genes
are (fibroblast,subclone 1) niche differential. We can do this using the
<code><a href="../reference/compute_significance.html">compute_significance()</a></code> function. We set side = 1 and
direction = “pos” so that the niche differential genes correspond to
those that are upregulated in fibroblasts when in the presence of tumor
subclone 1.</p>
<pre><code><span><span class="co">#get significant genes</span></span>
<span><span class="va">sig_genes</span> <span class="op">=</span> <span class="fu"><a href="../reference/compute_significance.html">compute_significance</a></span><span class="op">(</span>input_list <span class="op">=</span> <span class="va">res</span>,cell_type <span class="op">=</span> <span class="st">"stromal"</span>,</span>
<span>                                 effect_name <span class="op">=</span> <span class="st">"subclone_1"</span>,</span>
<span>                                 beta_name <span class="op">=</span> <span class="st">"beta_estimate"</span>,</span>
<span>                                 standard_error_name <span class="op">=</span> <span class="st">"standard_error_matrix"</span>,</span>
<span>                                 sided <span class="op">=</span> <span class="fl">1</span>,direction <span class="op">=</span> <span class="st">"pos"</span><span class="op">)</span></span>
<span><span class="va">sig_genes</span> <span class="op">=</span> <span class="va">sig_genes</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">pval</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"The following genes are (fibroblast,tumor subclone 1) niche differential"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sig_genes</span><span class="op">)</span><span class="op">[</span><span class="va">sig_genes</span><span class="op">$</span><span class="va">qval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="downstream-analysis-determining-niche-differential-marker-genes">Downstream Analysis: Determining Niche Differential Marker
Genes<a class="anchor" aria-label="anchor" href="#downstream-analysis-determining-niche-differential-marker-genes"></a>
</h4>
<p>We can also perform a contrast test to see if a gene is niche-marker
gene. A niche marker gene for a cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
is one in which
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>β</mi><mrow><msub><mi>t</mi><mn>1</mn></msub><mo>,</mo><mi>g</mi></mrow><mi>t</mi></msubsup><mo>≠</mo><msubsup><mi>β</mi><mrow><msub><mi>t</mi><mn>2</mn></msub><mo>,</mo><mi>g</mi></mrow><mi>t</mi></msubsup></mrow><annotation encoding="application/x-tex">\beta^{t}_{t_1,g} \neq \beta^{t}_{t_2,g}</annotation></semantics></math>
that is that we can differentiate whether cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>t</mi><annotation encoding="application/x-tex">t</annotation></semantics></math>
is near a niche enriched by cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>t</mi><mn>1</mn></msub><annotation encoding="application/x-tex">t_1</annotation></semantics></math>
or cell type
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>t</mi><mn>2</mn></msub><annotation encoding="application/x-tex">t_2</annotation></semantics></math>
by looking at its gene expression for gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>.
We can do this using the <code><a href="../reference/compute_contrast_significance.html">compute_contrast_significance()</a></code>
function. Here we are finding niche marker genes in fibroblasts when
near tumor subclone 1 vs tumor subclone 2. We set side = 1 and direction
= “pos” so that the genes correspond to those that are upregulated in
fibroblasts when in the presence of tumor subclone 1 vs when near tumor
subclone 2.</p>
<pre><code><span><span class="co">#compute contrast test</span></span>
<span><span class="va">sig_genes</span> <span class="op">=</span> <span class="fu"><a href="../reference/compute_contrast_significance.html">compute_contrast_significance</a></span><span class="op">(</span>input_list <span class="op">=</span> <span class="va">res</span>,</span>
<span>                                    cell_type <span class="op">=</span> <span class="st">"stromal"</span>,</span>
<span>                                    effect_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"subclone_1"</span>,<span class="st">"subclone_2"</span><span class="op">)</span>,</span>
<span>                               beta_name <span class="op">=</span> <span class="st">"beta_estimate"</span>,covariance_name <span class="op">=</span> <span class="st">"vcov"</span>,</span>
<span>                               sided <span class="op">=</span> <span class="fl">1</span>,direction <span class="op">=</span> <span class="st">"pos"</span><span class="op">)</span></span>
<span><span class="va">sig_genes</span> <span class="op">=</span> <span class="va">sig_genes</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">pval</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"The following genes are niche marker genes for fibroblasts near"</span>,</span>
<span>            <span class="st">" tumor sublcone 1 vs tumor subclone 2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sig_genes</span><span class="op">)</span><span class="op">[</span><span class="va">sig_genes</span><span class="op">$</span><span class="va">qval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="plots">Plots<a class="anchor" aria-label="anchor" href="#plots"></a>
</h4>
<p>Note that ISG15 is both a (fibroblast,subclone 1) niche differential
gene as well as a niche marker gene. If we plot the expression of ISG15
we see this visually. The fibroblasts in the subclone 2 niche express
much less ISG15 than that in the subclone 1 niche. Additionally, the non
tumor region (“other) shows very little expression of ISG15.</p>
<pre><code><span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="st">"ISG15"</span></span>
<span></span>
<span><span class="co"># Expression clipping</span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span>, <span class="va">gene</span><span class="op">]</span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmax</a></span><span class="op">(</span><span class="va">expression</span>, <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">expression</span>, <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmin</a></span><span class="op">(</span><span class="va">expression</span>, <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">expression</span>, <span class="fl">0.95</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#deconvolution clipping</span></span>
<span><span class="va">deconv_prop</span> <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">[</span>, <span class="st">"stromal"</span><span class="op">]</span></span>
<span><span class="va">deconv_prop</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmax</a></span><span class="op">(</span><span class="va">deconv_prop</span> , <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">deconv_prop</span> , <span class="fl">0.01</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">deconv_prop</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmin</a></span><span class="op">(</span><span class="va">deconv_prop</span> , <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">deconv_prop</span> , <span class="fl">0.99</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># === Plot 1: ISG15 Expression ===</span></span>
<span><span class="va">df1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  gene <span class="op">=</span> <span class="va">expression</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df1</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">gene</span>, <span class="st">" Expression"</span><span class="op">)</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"X"</span>, y <span class="op">=</span> <span class="st">"Y"</span>, color <span class="op">=</span> <span class="va">gene</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># === Plot 2: Majority subclones ===</span></span>
<span><span class="va">majority_subclone</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/maxCol.html" class="external-link">max.col</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">deconv</span>,</span>
<span>                                                   ties.method <span class="op">=</span> <span class="st">"first"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">df2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  majority <span class="op">=</span> <span class="va">majority_subclone</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">df2</span><span class="op">$</span><span class="va">highlight</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span></span>
<span>  <span class="va">df2</span><span class="op">$</span><span class="va">majority</span> <span class="op">==</span> <span class="st">"subclone_1"</span> <span class="op">~</span> <span class="st">"subclone_1"</span>,</span>
<span>  <span class="va">df2</span><span class="op">$</span><span class="va">majority</span> <span class="op">==</span> <span class="st">"subclone_2"</span> <span class="op">~</span> <span class="st">"subclone_2"</span>,</span>
<span>  <span class="cn">TRUE</span> <span class="op">~</span> <span class="st">"other"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df2</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">highlight</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span></span>
<span>    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"subclone_1"</span> <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>      <span class="st">"subclone_2"</span> <span class="op">=</span> <span class="st">"blue"</span>,</span>
<span>      <span class="st">"other"</span> <span class="op">=</span> <span class="st">"grey80"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Majority Subclone Regions"</span>, color <span class="op">=</span> <span class="st">"Majority Subclone"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># === Plot 3: Stromal proportions ===</span></span>
<span><span class="va">df3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  stromal <span class="op">=</span> <span class="va">deconv_prop</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df3</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">stromal</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Stromal Proportion"</span>, color <span class="op">=</span> <span class="st">"Stromal"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># === Combine all plots ===</span></span>
<span><span class="va">p1</span></span>
<span><span class="va">p2</span></span>
<span><span class="va">p3</span></span></code></pre>
<p>Note that COL4A1 is a (fibroblast,subclone 1) niche differential gene
but not niche marker gene. If we plot the expression of COL4A1 we see
this visually.</p>
<pre><code><span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="st">"COL4A1"</span></span>
<span></span>
<span><span class="co"># Expression clipping</span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span>, <span class="va">gene</span><span class="op">]</span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmax</a></span><span class="op">(</span><span class="va">expression</span>, <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">expression</span>, <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmin</a></span><span class="op">(</span><span class="va">expression</span>, <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">expression</span>, <span class="fl">0.95</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#deconvolution clipping</span></span>
<span><span class="va">deconv_prop</span> <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">[</span>, <span class="st">"stromal"</span><span class="op">]</span></span>
<span><span class="va">deconv_prop</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmax</a></span><span class="op">(</span><span class="va">deconv_prop</span> , <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">deconv_prop</span> , <span class="fl">0.01</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">deconv_prop</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmin</a></span><span class="op">(</span><span class="va">deconv_prop</span> , <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">deconv_prop</span> , <span class="fl">0.99</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># === Plot 1: ISG15 Expression ===</span></span>
<span><span class="va">df1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  gene <span class="op">=</span> <span class="va">expression</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df1</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">gene</span>, <span class="st">" Expression"</span><span class="op">)</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"X"</span>, y <span class="op">=</span> <span class="st">"Y"</span>, color <span class="op">=</span> <span class="va">gene</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># === Plot 2: Majority subclones ===</span></span>
<span><span class="va">majority_subclone</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">deconv</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/maxCol.html" class="external-link">max.col</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">deconv</span>,</span>
<span>                                                   ties.method <span class="op">=</span> <span class="st">"first"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">df2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  majority <span class="op">=</span> <span class="va">majority_subclone</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">df2</span><span class="op">$</span><span class="va">highlight</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span></span>
<span>  <span class="va">df2</span><span class="op">$</span><span class="va">majority</span> <span class="op">==</span> <span class="st">"subclone_1"</span> <span class="op">~</span> <span class="st">"subclone_1"</span>,</span>
<span>  <span class="va">df2</span><span class="op">$</span><span class="va">majority</span> <span class="op">==</span> <span class="st">"subclone_2"</span> <span class="op">~</span> <span class="st">"subclone_2"</span>,</span>
<span>  <span class="cn">TRUE</span> <span class="op">~</span> <span class="st">"other"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df2</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">highlight</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span></span>
<span>    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"subclone_1"</span> <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>      <span class="st">"subclone_2"</span> <span class="op">=</span> <span class="st">"blue"</span>,</span>
<span>      <span class="st">"other"</span> <span class="op">=</span> <span class="st">"grey80"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Majority Subclone Regions"</span>, color <span class="op">=</span> <span class="st">"Majority Subclone"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># === Plot 3: Stromal proportions ===</span></span>
<span><span class="va">df3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  y <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">coords</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>,</span>
<span>  stromal <span class="op">=</span> <span class="va">deconv_prop</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df3</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, color <span class="op">=</span> <span class="va">stromal</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Stromal Proportion"</span>, color <span class="op">=</span> <span class="st">"Stromal"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># === Combine all plots ===</span></span>
<span><span class="va">p1</span></span>
<span><span class="va">p2</span></span>
<span><span class="va">p3</span></span></code></pre>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kaishu Mason, Yijia Jiang, Nancy R. Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
